

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>10x scRNA-seq gene expression data (part 1) &#8212; Allen Brain Cell Atlas - Data Access</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/10x_snRNASeq_tutorial_part_1';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/allen_logo.jpeg" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../_static/allen_logo.jpeg" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Allen Brain Cell Atlas
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="getting_started.html">Getting started</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10Xv2.html">Whole mouse brain 10Xv2 single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10Xv3.html">Whole mouse brain 10Xv3 single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10X.html">Clustering analysis of whole adult mouse brain 10X single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-taxonomy.html">Whole adult mouse brain taxonomy of cell types</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/MERFISH-C57BL6J-638850.html">MERFISH spatial transcriptomics dataset of a single adult mouse brain</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-neighborhoods.html">Cluster groups and neighborhood specific UMAP embeddings</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/Allen-CCF-2020.html">Allen Mouse Common Coordinate Framework (2020 version)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/MERFISH-C57BL6J-638850-CCF.html">MERFISH CCF mapped coordinates</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cluster_annotation_tutorial.html">10x scRNA-seq clustering analysis and annotation (CCN20230504)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_groups_and_embeddings_tutorial.html">Cluster neighborhoods and embeddings</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_neighborhood_gallery.html">Cell type neighborhood gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="ccf_and_parcellation_annotation_tutorial.html">Allen Mouse Common Coordinate Framework (2020 version)</a></li>
<li class="toctree-l1"><a class="reference internal" href="merfish_ccf_registration_tutorial.html">MERFISH CCF mapped coordinates</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fnotebooks/10x_snRNASeq_tutorial_part_1.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/10x_snRNASeq_tutorial_part_1.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>10x scRNA-seq gene expression data (part 1)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-overview">Data overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-metadata">Cell metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#umap-spatial-embedding">UMAP spatial embedding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-transcriptomes">Single cell transcriptomes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-expression-matrices">Gene expression matrices</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-use-cases">Example use cases</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-of-canonical-neurotransmitter-transporter-genes-in-the-thalamus">Expression of canonical neurotransmitter transporter genes in the thalamus</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-of-tachykinin-2-tac2-in-the-thalamus">Expression of Tachykinin 2 (Tac2) in the thalamus</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <div class="tex2jax_ignore mathjax_ignore section" id="x-scrna-seq-gene-expression-data-part-1">
<h1>10x scRNA-seq gene expression data (part 1)<a class="headerlink" href="#x-scrna-seq-gene-expression-data-part-1" title="Permalink to this headline">#</a></h1>
<p>To efficiently handle data transfer and download, the 4 million cell 10X dataset, was subdivide into 23 expression matrices where cells are grouped by 10X chemistry (10Xv2 vs 10Xv3) and the broad anatomical region from where the cells were dissected from.</p>
<p>The purpose of this set of notebooks is to provide an overview of the data, the file organization and how to combine data and metadata through example use cases.</p>
<p>You need to be connected to the internet to run this notebook and that you have downloaded the example data via the getting started notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">requests</span>
</pre></div>
</div>
</div>
</div>
<p>The prerequisite for running this notebook is that the data have been downloaded to local directory maintaining the organization from the manifest.json. <strong>Change the download_base variable to where you have downloaded the data in your system.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">download_base</span> <span class="o">=</span> <span class="s1">&#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com&#39;</span>

<span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/releases/20230630/manifest.json&#39;</span>
<span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
    
<span class="n">metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;WMB-10X&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">view_directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> 
                               <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;directory_listing&#39;</span><span class="p">][</span><span class="s1">&#39;WMB-10X&#39;</span><span class="p">][</span><span class="s1">&#39;directories&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">],</span> 
                              <span class="s1">&#39;views&#39;</span><span class="p">)</span>
<span class="n">cache_views</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="data-overview">
<h2>Data overview<a class="headerlink" href="#data-overview" title="Permalink to this headline">#</a></h2>
<div class="section" id="cell-metadata">
<h3>Cell metadata<a class="headerlink" href="#cell-metadata" title="Permalink to this headline">#</a></h3>
<p>Essential cell metadata is stored as a dataframe. Each row represents one cell indexed by a <em>cell label</em>. The <em>cell label</em> is the concatenation of barcode and name of the sample. In this context, the sample is the <em>barcoded cell sample</em> that represents a single load into one port of the 10x Chromium. Note that cell barcodes are only unique within a single <em>barcoded cell sample</em> and that the same barcode can be reused. The <em>barcoded cell sample</em> label or name is unique in the database.</p>
<p>Each cell is associated with a <em>library label</em>, <em>library method</em>, <em>donor label</em>, <em>donor genotype</em>, <em>donor sex</em>, dissection *region of interest acronym”, the corresponding coarse <em>anatomical division label</em> and the <em>matrix_prefix</em> identifying which data package this cell is part of.</p>
<p>Further, each cell is associated with a <em>cluster alias</em> representing which cluster this cell is a member of and (x,y) coordinates of the all cells UMAP in Figure 1 of the manuscript.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;cell_metadata&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">cell</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;cell_label&#39;</span><span class="p">:</span><span class="nb">str</span><span class="p">})</span>
<span class="n">cell</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cell_label&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of cells = &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell</span><span class="p">))</span>
<span class="n">cell</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of cells =  4057701
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library_label</th>
      <th>anatomical_division_label</th>
      <th>cluster_alias</th>
      <th>library_method</th>
      <th>region_of_interest_acronym</th>
      <th>donor_label</th>
      <th>donor_genotype</th>
      <th>donor_sex</th>
      <th>dataset_label</th>
      <th>matrix_label</th>
      <th>x</th>
      <th>y</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GACTGCGAGCTGCAAG-020_C01</th>
      <td>L8TX_180406_01_F01</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381293</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>M</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>-6.997189</td>
      <td>-10.805128</td>
    </tr>
    <tr>
      <th>CGATGTAGTGCTAGCC-021_D01</th>
      <td>L8TX_180406_01_C02</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381295</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>F</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>-6.526481</td>
      <td>-10.935798</td>
    </tr>
    <tr>
      <th>CACATAGCATGGTTGT-020_D01</th>
      <td>L8TX_180406_01_B02</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381293</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>M</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>-6.673367</td>
      <td>-10.980438</td>
    </tr>
    <tr>
      <th>CCTAAAGGTGACCAAG-020_C01</th>
      <td>L8TX_180406_01_F01</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381293</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>M</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>-6.980066</td>
      <td>-10.790136</td>
    </tr>
    <tr>
      <th>CGGAGCTAGCTTTGGT-021_D01</th>
      <td>L8TX_180406_01_C02</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381295</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>F</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>-6.490485</td>
      <td>-10.884357</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can use panadas groupby function to see how many unique items are associated for each field and list them out if the number of items is small.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">print_column_info</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="p">:</span>
    
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="p">:</span>
        <span class="n">grouped</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="n">c</span><span class="p">]]</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
        <span class="n">members</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">grouped</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">30</span> <span class="p">:</span>
            <span class="n">members</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">grouped</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of unique </span><span class="si">%s</span><span class="s2"> = </span><span class="si">%d</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">grouped</span><span class="p">),</span><span class="n">members</span><span class="p">))</span>
        
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_column_info</span><span class="p">(</span><span class="n">cell</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique library_label = 781 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique anatomical_division_label = 12 [&#39;CB&#39;, &#39;CTXsp&#39;, &#39;HPF&#39;, &#39;HY&#39;, &#39;Isocortex&#39;, &#39;MB&#39;, &#39;MY&#39;, &#39;OLF&#39;, &#39;P&#39;, &#39;PAL&#39;, &#39;STR&#39;, &#39;TH&#39;]
Number of unique cluster_alias = 5196 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique library_method = 2 [&#39;10Xv2&#39;, &#39;10Xv3&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique region_of_interest_acronym = 29 [&#39;ACA&#39;, &#39;AI&#39;, &#39;AUD&#39;, &#39;AUD-TEa-PERI-ECT&#39;, &#39;CB&#39;, &#39;CTXsp&#39;, &#39;ENT&#39;, &#39;HIP&#39;, &#39;HY&#39;, &#39;LSX&#39;, &#39;MB&#39;, &#39;MO-FRP&#39;, &#39;MOp&#39;, &#39;MY&#39;, &#39;OLF&#39;, &#39;P&#39;, &#39;PAL&#39;, &#39;PL-ILA-ORB&#39;, &#39;RHP&#39;, &#39;RSP&#39;, &#39;SS-GU-VISC&#39;, &#39;SSp&#39;, &#39;STRd&#39;, &#39;STRv&#39;, &#39;TEa-PERI-ECT&#39;, &#39;TH&#39;, &#39;VIS&#39;, &#39;VIS-PTLp&#39;, &#39;sAMY&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique donor_label = 318 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique donor_genotype = 4 [&#39;Ai14(RCL-tdT)/wt&#39;, &#39;Gad2-IRES-Cre/wt;Ai14(RCL-tdT)/wt&#39;, &#39;Slc32a1-IRES-Cre/wt;Ai14(RCL-tdT)/wt&#39;, &#39;Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique donor_sex = 2 [&#39;F&#39;, &#39;M&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique dataset_label = 2 [&#39;WMB-10Xv2&#39;, &#39;WMB-10Xv3&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique matrix_label = 23 [&#39;WMB-10Xv2-CTXsp&#39;, &#39;WMB-10Xv2-HPF&#39;, &#39;WMB-10Xv2-HY&#39;, &#39;WMB-10Xv2-Isocortex-1&#39;, &#39;WMB-10Xv2-Isocortex-2&#39;, &#39;WMB-10Xv2-Isocortex-3&#39;, &#39;WMB-10Xv2-Isocortex-4&#39;, &#39;WMB-10Xv2-MB&#39;, &#39;WMB-10Xv2-OLF&#39;, &#39;WMB-10Xv2-TH&#39;, &#39;WMB-10Xv3-CB&#39;, &#39;WMB-10Xv3-CTXsp&#39;, &#39;WMB-10Xv3-HPF&#39;, &#39;WMB-10Xv3-HY&#39;, &#39;WMB-10Xv3-Isocortex-1&#39;, &#39;WMB-10Xv3-Isocortex-2&#39;, &#39;WMB-10Xv3-MB&#39;, &#39;WMB-10Xv3-MY&#39;, &#39;WMB-10Xv3-OLF&#39;, &#39;WMB-10Xv3-P&#39;, &#39;WMB-10Xv3-PAL&#39;, &#39;WMB-10Xv3-STR&#39;, &#39;WMB-10Xv3-TH&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique x = 4021254 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique y = 4022997 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;library_method&#39;</span><span class="p">)[[</span><span class="s1">&#39;x&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x</th>
    </tr>
    <tr>
      <th>library_method</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10Xv2</th>
      <td>1708157</td>
    </tr>
    <tr>
      <th>10Xv3</th>
      <td>2349544</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can also bring in the pivot table from the “cluster annotation tutorial” to associate each cell with terms at each cell type classification level.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">taxonomy_metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;WMB-taxonomy&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">taxonomy_metadata</span><span class="p">[</span><span class="s1">&#39;cluster_to_cluster_annotation_membership_pivoted&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">cluster_details</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">cluster_details</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cluster_details</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>neurotransmitter</th>
      <th>division</th>
      <th>class</th>
      <th>subclass</th>
      <th>supertype</th>
      <th>cluster</th>
    </tr>
    <tr>
      <th>cluster_alias</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0328 L2 IT PPP-APr Glut_3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0080 L2 IT PPP-APr Glut_2</td>
      <td>0323 L2 IT PPP-APr Glut_2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0080 L2 IT PPP-APr Glut_2</td>
      <td>0324 L2 IT PPP-APr Glut_2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0080 L2 IT PPP-APr Glut_2</td>
      <td>0326 L2 IT PPP-APr Glut_2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">taxonomy_metadata</span><span class="p">[</span><span class="s1">&#39;cluster_to_cluster_annotation_membership_color&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">cluster_colors</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">cluster_colors</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cluster_colors</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>neurotransmitter_color</th>
      <th>division_color</th>
      <th>class_color</th>
      <th>subclass_color</th>
      <th>supertype_color</th>
      <th>cluster_color</th>
    </tr>
    <tr>
      <th>cluster_alias</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
    </tr>
    <tr>
      <th>2</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#280099</td>
    </tr>
    <tr>
      <th>3</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#002BCC</td>
      <td>#5CCCA0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#002BCC</td>
      <td>#CC6D1F</td>
    </tr>
    <tr>
      <th>5</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#002BCC</td>
      <td>#FF8E26</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;region_of_interest_metadata&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">roi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">roi</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;acronym&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">roi</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;order&#39;</span><span class="p">:</span><span class="s1">&#39;region_of_interest_order&#39;</span><span class="p">,</span><span class="s1">&#39;color_hex_triplet&#39;</span><span class="p">:</span><span class="s1">&#39;region_of_interest_color&#39;</span><span class="p">},</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">roi</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>label</th>
      <th>name</th>
      <th>region_of_interest_order</th>
      <th>region_of_interest_color</th>
    </tr>
    <tr>
      <th>acronym</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>MO-FRP</th>
      <td>WMB-MO-FRP</td>
      <td>Somatomotor - Frontal pole</td>
      <td>0</td>
      <td>#3DCC7C</td>
    </tr>
    <tr>
      <th>MOp</th>
      <td>WMB-MOp</td>
      <td>Primary motor area</td>
      <td>1</td>
      <td>#179968</td>
    </tr>
    <tr>
      <th>SS-GU-VISC</th>
      <td>WMB-SS-GU-VISC</td>
      <td>Somatosensory/gustatory/visceral areas</td>
      <td>2</td>
      <td>#2E8599</td>
    </tr>
    <tr>
      <th>SSp</th>
      <td>WMB-SSp</td>
      <td>Primary somatosensory area</td>
      <td>3</td>
      <td>#5CCCCC</td>
    </tr>
    <tr>
      <th>AUD</th>
      <td>WMB-AUD</td>
      <td>Auditory areas</td>
      <td>4</td>
      <td>#455A99</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_extended</span> <span class="o">=</span> <span class="n">cell</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cluster_details</span><span class="p">,</span><span class="n">on</span><span class="o">=</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">)</span>
<span class="n">cell_extended</span> <span class="o">=</span> <span class="n">cell_extended</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cluster_colors</span><span class="p">,</span><span class="n">on</span><span class="o">=</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">)</span>
<span class="n">cell_extended</span> <span class="o">=</span> <span class="n">cell_extended</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">roi</span><span class="p">[[</span><span class="s1">&#39;region_of_interest_order&#39;</span><span class="p">,</span><span class="s1">&#39;region_of_interest_color&#39;</span><span class="p">]],</span><span class="n">on</span><span class="o">=</span><span class="s1">&#39;region_of_interest_acronym&#39;</span><span class="p">)</span>
<span class="n">cell_extended</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library_label</th>
      <th>anatomical_division_label</th>
      <th>cluster_alias</th>
      <th>library_method</th>
      <th>region_of_interest_acronym</th>
      <th>donor_label</th>
      <th>donor_genotype</th>
      <th>donor_sex</th>
      <th>dataset_label</th>
      <th>matrix_label</th>
      <th>...</th>
      <th>supertype</th>
      <th>cluster</th>
      <th>neurotransmitter_color</th>
      <th>division_color</th>
      <th>class_color</th>
      <th>subclass_color</th>
      <th>supertype_color</th>
      <th>cluster_color</th>
      <th>region_of_interest_order</th>
      <th>region_of_interest_color</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GACTGCGAGCTGCAAG-020_C01</th>
      <td>L8TX_180406_01_F01</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381293</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>M</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>...</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
      <td>15</td>
      <td>#CCB05C</td>
    </tr>
    <tr>
      <th>CGATGTAGTGCTAGCC-021_D01</th>
      <td>L8TX_180406_01_C02</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381295</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>F</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>...</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
      <td>15</td>
      <td>#CCB05C</td>
    </tr>
    <tr>
      <th>CACATAGCATGGTTGT-020_D01</th>
      <td>L8TX_180406_01_B02</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381293</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>M</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>...</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
      <td>15</td>
      <td>#CCB05C</td>
    </tr>
    <tr>
      <th>CCTAAAGGTGACCAAG-020_C01</th>
      <td>L8TX_180406_01_F01</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381293</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>M</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>...</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
      <td>15</td>
      <td>#CCB05C</td>
    </tr>
    <tr>
      <th>CGGAGCTAGCTTTGGT-021_D01</th>
      <td>L8TX_180406_01_C02</td>
      <td>HPF</td>
      <td>1</td>
      <td>10Xv2</td>
      <td>RHP</td>
      <td>Snap25-IRES2-Cre;Ai14-381295</td>
      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
      <td>F</td>
      <td>WMB-10Xv2</td>
      <td>WMB-10Xv2-HPF</td>
      <td>...</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
      <td>15</td>
      <td>#CCB05C</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 26 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_column_info</span><span class="p">(</span><span class="n">cell_extended</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique library_label = 781 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique anatomical_division_label = 12 [&#39;CB&#39;, &#39;CTXsp&#39;, &#39;HPF&#39;, &#39;HY&#39;, &#39;Isocortex&#39;, &#39;MB&#39;, &#39;MY&#39;, &#39;OLF&#39;, &#39;P&#39;, &#39;PAL&#39;, &#39;STR&#39;, &#39;TH&#39;]
Number of unique cluster_alias = 5196 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique library_method = 2 [&#39;10Xv2&#39;, &#39;10Xv3&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique region_of_interest_acronym = 29 [&#39;ACA&#39;, &#39;AI&#39;, &#39;AUD&#39;, &#39;AUD-TEa-PERI-ECT&#39;, &#39;CB&#39;, &#39;CTXsp&#39;, &#39;ENT&#39;, &#39;HIP&#39;, &#39;HY&#39;, &#39;LSX&#39;, &#39;MB&#39;, &#39;MO-FRP&#39;, &#39;MOp&#39;, &#39;MY&#39;, &#39;OLF&#39;, &#39;P&#39;, &#39;PAL&#39;, &#39;PL-ILA-ORB&#39;, &#39;RHP&#39;, &#39;RSP&#39;, &#39;SS-GU-VISC&#39;, &#39;SSp&#39;, &#39;STRd&#39;, &#39;STRv&#39;, &#39;TEa-PERI-ECT&#39;, &#39;TH&#39;, &#39;VIS&#39;, &#39;VIS-PTLp&#39;, &#39;sAMY&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique donor_label = 318 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique donor_genotype = 4 [&#39;Ai14(RCL-tdT)/wt&#39;, &#39;Gad2-IRES-Cre/wt;Ai14(RCL-tdT)/wt&#39;, &#39;Slc32a1-IRES-Cre/wt;Ai14(RCL-tdT)/wt&#39;, &#39;Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique donor_sex = 2 [&#39;F&#39;, &#39;M&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique dataset_label = 2 [&#39;WMB-10Xv2&#39;, &#39;WMB-10Xv3&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique matrix_label = 23 [&#39;WMB-10Xv2-CTXsp&#39;, &#39;WMB-10Xv2-HPF&#39;, &#39;WMB-10Xv2-HY&#39;, &#39;WMB-10Xv2-Isocortex-1&#39;, &#39;WMB-10Xv2-Isocortex-2&#39;, &#39;WMB-10Xv2-Isocortex-3&#39;, &#39;WMB-10Xv2-Isocortex-4&#39;, &#39;WMB-10Xv2-MB&#39;, &#39;WMB-10Xv2-OLF&#39;, &#39;WMB-10Xv2-TH&#39;, &#39;WMB-10Xv3-CB&#39;, &#39;WMB-10Xv3-CTXsp&#39;, &#39;WMB-10Xv3-HPF&#39;, &#39;WMB-10Xv3-HY&#39;, &#39;WMB-10Xv3-Isocortex-1&#39;, &#39;WMB-10Xv3-Isocortex-2&#39;, &#39;WMB-10Xv3-MB&#39;, &#39;WMB-10Xv3-MY&#39;, &#39;WMB-10Xv3-OLF&#39;, &#39;WMB-10Xv3-P&#39;, &#39;WMB-10Xv3-PAL&#39;, &#39;WMB-10Xv3-STR&#39;, &#39;WMB-10Xv3-TH&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique x = 4021254 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique y = 4022997 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique neurotransmitter = 9 [&#39;Chol&#39;, &#39;Dopa&#39;, &#39;GABA&#39;, &#39;GABA-Glyc&#39;, &#39;Glut&#39;, &#39;Glut-GABA&#39;, &#39;Hist&#39;, &#39;Nora&#39;, &#39;Sero&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique division = 7 [&#39;1 Pallium glutamatergic&#39;, &#39;2 Subpallium GABAergic&#39;, &#39;3 PAL-sAMY-TH-HY-MB-HB neuronal&#39;, &#39;4 CBX-MOB-other neuronal&#39;, &#39;5 Neuroglial&#39;, &#39;6 Vascular&#39;, &#39;7 Immune&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique class = 32 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique subclass = 307 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique supertype = 1115 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique cluster = 5196 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique neurotransmitter_color = 10 [&#39;#03EDFF&#39;, &#39;#0a9964&#39;, &#39;#2B93DF&#39;, &#39;#533691&#39;, &#39;#73E785&#39;, &#39;#820e57&#39;, &#39;#FF3358&#39;, &#39;#f9f9f9&#39;, &#39;#fcf04b&#39;, &#39;#ff7621&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique division_color = 7 [&#39;#4D58FF&#39;, &#39;#53776C&#39;, &#39;#94AF97&#39;, &#39;#9F3DCC&#39;, &#39;#A6906F&#39;, &#39;#FF2678&#39;, &#39;#fc810f&#39;]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique class_color = 32 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique subclass_color = 307 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique supertype_color = 1115 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique cluster_color = 5195 
Number of unique region_of_interest_order = 29 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of unique region_of_interest_color = 29 [&#39;#0059CC&#39;, &#39;#00990A&#39;, &#39;#00FFCF&#39;, &#39;#179968&#39;, &#39;#26AEFF&#39;, &#39;#2E8599&#39;, &#39;#300099&#39;, &#39;#341FCC&#39;, &#39;#34CC1F&#39;, &#39;#3DCC7C&#39;, &#39;#455A99&#39;, &#39;#4D58FF&#39;, &#39;#5CCCCC&#39;, &#39;#6F9945&#39;, &#39;#84FF4D&#39;, &#39;#8CCC00&#39;, &#39;#992E6A&#39;, &#39;#994817&#39;, &#39;#99922E&#39;, &#39;#9F3DCC&#39;, &#39;#B973FF&#39;, &#39;#CC0026&#39;, &#39;#CC5CB0&#39;, &#39;#CCB05C&#39;, &#39;#E4FF26&#39;, &#39;#FF00EF&#39;, &#39;#FF2678&#39;, &#39;#FF7373&#39;, &#39;#FF8F00&#39;]
</pre></div>
</div>
</div>
</div>
<p>For convenience, we can cache this view for later reuse.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="s1">&#39;cell_metadata_with_cluster_annotation.csv&#39;</span><span class="p">)</span>
    <span class="n">cell_extended</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="umap-spatial-embedding">
<h3>UMAP spatial embedding<a class="headerlink" href="#umap-spatial-embedding" title="Permalink to this headline">#</a></h3>
<p>Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique that can be used for visualizing and exploring large-dimension datasets. The x, y columns of the cell metadata table represents the coordinate of the all cells UMAP in Figure 1 of the manuscript.</p>
<p>We define a small helper function <em>plot umap</em> to visualize the cells on the UMAP colorized by: dissection region of interest, neurotransmitter identity, cell types division, class and subclass. For ease of demostration, we do a simple subsampling of the cells by a factor of 10 to reduce processing time.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_umap</span><span class="p">(</span> <span class="n">xx</span><span class="p">,</span> <span class="n">yy</span><span class="p">,</span> <span class="n">cc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">val</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fig_width</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span> <span class="p">)</span> <span class="p">:</span>
    
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="n">fig_width</span><span class="p">,</span> <span class="n">fig_height</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">cmap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">xx</span><span class="p">,</span><span class="n">yy</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">val</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">cc</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">xx</span><span class="p">,</span><span class="n">yy</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">cc</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
        
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">18</span><span class="p">,</span><span class="mi">27</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mi">18</span><span class="p">,</span><span class="mi">27</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
    
    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_subsampled</span> <span class="o">=</span> <span class="n">cell_extended</span><span class="o">.</span><span class="n">loc</span><span class="p">[::</span><span class="mi">10</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cell_subsampled</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>405771
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_umap</span><span class="p">(</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">cc</span><span class="o">=</span><span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;region_of_interest_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Dissection Region Of Interest&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/741a7b9931fd0f3d39c570e664f1b4dfadde23ba91c061f3f8ab587f00b6542f.png" src="../_images/741a7b9931fd0f3d39c570e664f1b4dfadde23ba91c061f3f8ab587f00b6542f.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_umap</span><span class="p">(</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">cc</span><span class="o">=</span><span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;neurotransmitter_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Neuortransmitter Identity&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1355041c711e08649e7c060c4587e1b3daa6cae68f332dffac48ed39fe032885.png" src="../_images/1355041c711e08649e7c060c4587e1b3daa6cae68f332dffac48ed39fe032885.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_umap</span><span class="p">(</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">cc</span><span class="o">=</span><span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;division_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell Types: Division&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/047aba3272b361093fd049f5822233a265039d529152e099aabf44fd44da743d.png" src="../_images/047aba3272b361093fd049f5822233a265039d529152e099aabf44fd44da743d.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_umap</span><span class="p">(</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">cc</span><span class="o">=</span><span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;class_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell Types: Class&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/2d15038b8b92965c895c2c61c6d40a933ce4f29cc0a14c350f7597c53d3a376c.png" src="../_images/2d15038b8b92965c895c2c61c6d40a933ce4f29cc0a14c350f7597c53d3a376c.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_umap</span><span class="p">(</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span> <span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">cc</span><span class="o">=</span><span class="n">cell_subsampled</span><span class="p">[</span><span class="s1">&#39;subclass_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell Types: Subclass&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/62492b0d28388815657af0ce5f9842cbd53fe95880d586558cdecf6bd2d09b6c.png" src="../_images/62492b0d28388815657af0ce5f9842cbd53fe95880d586558cdecf6bd2d09b6c.png" />
</div>
</div>
</div>
<div class="section" id="single-cell-transcriptomes">
<h3>Single cell transcriptomes<a class="headerlink" href="#single-cell-transcriptomes" title="Permalink to this headline">#</a></h3>
<p>Each 10x library was sequenced on the Illumina NovaSeq6000 platform, and sequencing reads were aligned to the mouse
reference transcriptome mm10 (GENCODE vM23/Ensembl 98) using the 10x Genomics CellRanger pipeline (version 6.1.1) with default parameters. Each gene is uniquely identifier by an Ensembl ID. It is best practice to gene identifier to for tracking and data interchange as gene symbols are not unique and can change over time</p>
<p>Each row of the gene dataframe has Ensembl gene identifier, a gene symbol and name. 1173 genes have no expression across all the cells in the dataset and can be identified by “no expression” in the comment column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">gene</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene_identifier&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of genes = &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
<span class="n">gene</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of genes =  32285
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
      <th>name</th>
      <th>comment</th>
    </tr>
    <tr>
      <th>gene_identifier</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSMUSG00000051951</th>
      <td>Xkr4</td>
      <td>X-linked Kx blood group related 4</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000089699</th>
      <td>Gm1992</td>
      <td>predicted gene 1992</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000102331</th>
      <td>Gm19938</td>
      <td>predicted gene, 19938</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000102343</th>
      <td>Gm37381</td>
      <td>predicted gene, 37381</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000025900</th>
      <td>Rp1</td>
      <td>retinitis pigmentosa 1 (human)</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;comment&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;no expression&#39;</span><span class="p">)</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of no expression genes&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered</span><span class="p">))</span>
<span class="n">filtered</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of no expression genes 1173
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
      <th>name</th>
      <th>comment</th>
    </tr>
    <tr>
      <th>gene_identifier</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSMUSG00000118272</th>
      <td>Arhgef4</td>
      <td>Rho guanine nucleotide exchange factor (GEF) 4</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000117809</th>
      <td>Asdurf</td>
      <td>Asnsd1 upstream reading frame</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000047383</th>
      <td>AC169382.1</td>
      <td>novel transcript</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000097001</th>
      <td>1700054K02Rik</td>
      <td>RIKEN cDNA 1700054K02 gene</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000116048</th>
      <td>Sept2</td>
      <td>septin 2</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>ENSMUSG00000091585</th>
      <td>AC140365.1</td>
      <td>PRAME family member 8-like</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000095523</th>
      <td>AC124606.1</td>
      <td>PRAME family member 8-like</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000095475</th>
      <td>AC133095.2</td>
      <td>uncharacterized LOC545763</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000094855</th>
      <td>AC133095.1</td>
      <td>uncharacterized LOC620639</td>
      <td>no expression</td>
    </tr>
    <tr>
      <th>ENSMUSG00000095019</th>
      <td>AC234645.1</td>
      <td>NaN</td>
      <td>no expression</td>
    </tr>
  </tbody>
</table>
<p>1173 rows × 3 columns</p>
</div></div></div>
</div>
</div>
<div class="section" id="gene-expression-matrices">
<h3>Gene expression matrices<a class="headerlink" href="#gene-expression-matrices" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">gene</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene_identifier&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of genes = &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
<span class="n">gene</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of genes =  32285
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
      <th>name</th>
      <th>comment</th>
    </tr>
    <tr>
      <th>gene_identifier</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSMUSG00000051951</th>
      <td>Xkr4</td>
      <td>X-linked Kx blood group related 4</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000089699</th>
      <td>Gm1992</td>
      <td>predicted gene 1992</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000102331</th>
      <td>Gm19938</td>
      <td>predicted gene, 19938</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000102343</th>
      <td>Gm37381</td>
      <td>predicted gene, 37381</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSMUSG00000025900</th>
      <td>Rp1</td>
      <td>retinitis pigmentosa 1 (human)</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The 4 million cell dataset has been divided into 23 expression matrices to make data transfer and download more efficient. Each package is formatted as an annadata h5ad file with minimal metadata. In this next section, we provide example code on how to open one of the expression matrix files and connect with the rich cell level metadata discussed above.</p>
<p>For each subset, there are two h5ad files one storing the raw counts and the other log normalization of it. For this example, we will use the log2 version of the expression matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expression_matrices</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;WMB-10Xv2&#39;</span><span class="p">][</span><span class="s1">&#39;expression_matrices&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expression_matrices</span><span class="p">[</span><span class="s1">&#39;WMB-10Xv2-TH&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;log2&#39;: {&#39;files&#39;: {&#39;h5ad&#39;: {&#39;version&#39;: &#39;20230630&#39;,
    &#39;relative_path&#39;: &#39;expression_matrices/WMB-10Xv2/20230630/WMB-10Xv2-TH-log2.h5ad&#39;,
    &#39;url&#39;: &#39;https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/expression_matrices/WMB-10Xv2/20230630/WMB-10Xv2-TH-log2.h5ad&#39;,
    &#39;size&#39;: 4028273658}}},
 &#39;raw&#39;: {&#39;files&#39;: {&#39;h5ad&#39;: {&#39;version&#39;: &#39;20230630&#39;,
    &#39;relative_path&#39;: &#39;expression_matrices/WMB-10Xv2/20230630/WMB-10Xv2-TH-raw.h5ad&#39;,
    &#39;url&#39;: &#39;https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/expression_matrices/WMB-10Xv2/20230630/WMB-10Xv2-TH-raw.h5ad&#39;,
    &#39;size&#39;: 4028273658}}}}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">expression_matrices</span><span class="p">[</span><span class="s1">&#39;WMB-10Xv2-TH&#39;</span><span class="p">][</span><span class="s1">&#39;log2&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;h5ad&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/expression_matrices/WMB-10Xv2/20230630/WMB-10Xv2-TH-log2.h5ad
</pre></div>
</div>
</div>
</div>
<p>We use the anndata’s <em>read_h5ad</em> function to open the package for the log2 normalization file. The “backed=’r’” makes use of the lazy loading functionality to only load required data. By default, anndata will load the entire expression matrix in memory.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="n">backed</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FileNotFoundError</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">28</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="n">backed</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="nn">File ~/miniconda3/envs/jupyter_book/lib/python3.9/site-packages/anndata/_io/h5ad.py:197,</span> in <span class="ni">read_h5ad</span><span class="nt">(filename, backed, as_sparse, as_sparse_fmt, chunk_size)</span>
<span class="g g-Whitespace">    </span><span class="mi">195</span>         <span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;r+&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">196</span>     <span class="k">assert</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">}</span>
<span class="ne">--&gt; </span><span class="mi">197</span>     <span class="k">return</span> <span class="n">read_h5ad_backed</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">mode</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">199</span> <span class="k">if</span> <span class="n">as_sparse_fmt</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">,</span> <span class="n">sparse</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">200</span>     <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">201</span>         <span class="s2">&quot;Dense formats can only be read to CSR or CSC matrices at this time.&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">202</span>     <span class="p">)</span>

<span class="nn">File ~/miniconda3/envs/jupyter_book/lib/python3.9/site-packages/anndata/_io/h5ad.py:128,</span> in <span class="ni">read_h5ad_backed</span><span class="nt">(filename, mode)</span>
<span class="g g-Whitespace">    </span><span class="mi">125</span> <span class="k">def</span> <span class="nf">read_h5ad_backed</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">],</span> <span class="n">mode</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">AnnData</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">126</span>     <span class="n">d</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">,</span> <span class="n">filemode</span><span class="o">=</span><span class="n">mode</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">128</span>     <span class="n">f</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">mode</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">130</span>     <span class="n">attributes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;obsm&quot;</span><span class="p">,</span> <span class="s2">&quot;varm&quot;</span><span class="p">,</span> <span class="s2">&quot;obsp&quot;</span><span class="p">,</span> <span class="s2">&quot;varp&quot;</span><span class="p">,</span> <span class="s2">&quot;uns&quot;</span><span class="p">,</span> <span class="s2">&quot;layers&quot;</span><span class="p">]</span>
<span class="g g-Whitespace">    </span><span class="mi">131</span>     <span class="n">df_attributes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;obs&quot;</span><span class="p">,</span> <span class="s2">&quot;var&quot;</span><span class="p">]</span>

<span class="nn">File ~/miniconda3/envs/jupyter_book/lib/python3.9/site-packages/h5py/_hl/files.py:567,</span> in <span class="ni">File.__init__</span><span class="nt">(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, alignment_threshold, alignment_interval, meta_block_size, **kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">558</span>     <span class="n">fapl</span> <span class="o">=</span> <span class="n">make_fapl</span><span class="p">(</span><span class="n">driver</span><span class="p">,</span> <span class="n">libver</span><span class="p">,</span> <span class="n">rdcc_nslots</span><span class="p">,</span> <span class="n">rdcc_nbytes</span><span class="p">,</span> <span class="n">rdcc_w0</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">559</span>                      <span class="n">locking</span><span class="p">,</span> <span class="n">page_buf_size</span><span class="p">,</span> <span class="n">min_meta_keep</span><span class="p">,</span> <span class="n">min_raw_keep</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">560</span>                      <span class="n">alignment_threshold</span><span class="o">=</span><span class="n">alignment_threshold</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">561</span>                      <span class="n">alignment_interval</span><span class="o">=</span><span class="n">alignment_interval</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">562</span>                      <span class="n">meta_block_size</span><span class="o">=</span><span class="n">meta_block_size</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">563</span>                      <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">564</span>     <span class="n">fcpl</span> <span class="o">=</span> <span class="n">make_fcpl</span><span class="p">(</span><span class="n">track_order</span><span class="o">=</span><span class="n">track_order</span><span class="p">,</span> <span class="n">fs_strategy</span><span class="o">=</span><span class="n">fs_strategy</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">565</span>                      <span class="n">fs_persist</span><span class="o">=</span><span class="n">fs_persist</span><span class="p">,</span> <span class="n">fs_threshold</span><span class="o">=</span><span class="n">fs_threshold</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">566</span>                      <span class="n">fs_page_size</span><span class="o">=</span><span class="n">fs_page_size</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">567</span>     <span class="n">fid</span> <span class="o">=</span> <span class="n">make_fid</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">userblock_size</span><span class="p">,</span> <span class="n">fapl</span><span class="p">,</span> <span class="n">fcpl</span><span class="p">,</span> <span class="n">swmr</span><span class="o">=</span><span class="n">swmr</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">569</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">libver</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">570</span>     <span class="bp">self</span><span class="o">.</span><span class="n">_libver</span> <span class="o">=</span> <span class="n">libver</span>

<span class="nn">File ~/miniconda3/envs/jupyter_book/lib/python3.9/site-packages/h5py/_hl/files.py:231,</span> in <span class="ni">make_fid</span><span class="nt">(name, mode, userblock_size, fapl, fcpl, swmr)</span>
<span class="g g-Whitespace">    </span><span class="mi">229</span>     <span class="k">if</span> <span class="n">swmr</span> <span class="ow">and</span> <span class="n">swmr_support</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">230</span>         <span class="n">flags</span> <span class="o">|=</span> <span class="n">h5f</span><span class="o">.</span><span class="n">ACC_SWMR_READ</span>
<span class="ne">--&gt; </span><span class="mi">231</span>     <span class="n">fid</span> <span class="o">=</span> <span class="n">h5f</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">flags</span><span class="p">,</span> <span class="n">fapl</span><span class="o">=</span><span class="n">fapl</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">232</span> <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">&#39;r+&#39;</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">233</span>     <span class="n">fid</span> <span class="o">=</span> <span class="n">h5f</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">h5f</span><span class="o">.</span><span class="n">ACC_RDWR</span><span class="p">,</span> <span class="n">fapl</span><span class="o">=</span><span class="n">fapl</span><span class="p">)</span>

<span class="nn">File h5py/_objects.pyx:54,</span> in <span class="ni">h5py._objects.with_phil.wrapper</span><span class="nt">()</span>

<span class="nn">File h5py/_objects.pyx:55,</span> in <span class="ni">h5py._objects.with_phil.wrapper</span><span class="nt">()</span>

<span class="nn">File h5py/h5f.pyx:106,</span> in <span class="ni">h5py.h5f.open</span><span class="nt">()</span>

<span class="ne">FileNotFoundError</span>: [Errno 2] Unable to open file (unable to open file: name = &#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/expression_matrices/WMB-10Xv2/20230630/WMB-10Xv2-TH-log2.h5ad&#39;, errno = 2, error message = &#39;No such file or directory&#39;, flags = 0, o_flags = 0)
</pre></div>
</div>
</div>
</div>
<p>Genes are represented as “variables”. For this data, the <em>var</em> dataframe is indexed by the Ensembl gene identifier with one metadata column <em>gene symbol</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of genes = &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">))</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of genes =  32285
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;ENSMUSG00000051951&#39;, &#39;ENSMUSG00000089699&#39;, &#39;ENSMUSG00000102331&#39;,
       &#39;ENSMUSG00000102343&#39;, &#39;ENSMUSG00000025900&#39;],
      dtype=&#39;object&#39;, name=&#39;gene_identifier&#39;)
</pre></div>
</div>
</div>
</div>
<p>Cells are represented as “observations”. For this data, the <em>obs</em> dataframe is index by <em>cell label</em> with three minimal metadata columns <em>cell barcode</em>, <em>library label</em> and <em>anatomical division label</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of cells = &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">))</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of cells =  131212
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;CAGGTGCAGGCTAGCA-040_C01&#39;, &#39;TGCGCAGGTTGCGCAC-045_C01&#39;,
       &#39;CGATGTATCTTGCCGT-042_B01&#39;, &#39;GACTAACGTCCTCTTG-040_B01&#39;,
       &#39;GATCGTACAACTGCTA-040_B01&#39;],
      dtype=&#39;object&#39;, name=&#39;cell_label&#39;)
</pre></div>
</div>
</div>
</div>
<p>We can easily connect the cells in the package to the extended cell metadata with simple pandas filtering.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_filtered</span> <span class="o">=</span> <span class="n">cell_extended</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of cells = &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_filtered</span><span class="p">))</span>
<span class="n">print_column_info</span><span class="p">(</span> <span class="n">cell_filtered</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of cells =  131212
Number of unique library_label = 26 [&#39;L8TX_180815_01_A07&#39;, &#39;L8TX_180815_01_A08&#39;, &#39;L8TX_180815_01_B07&#39;, &#39;L8TX_180815_01_B08&#39;, &#39;L8TX_180815_01_C07&#39;, &#39;L8TX_180815_01_C08&#39;, &#39;L8TX_180815_01_D07&#39;, &#39;L8TX_180815_01_D08&#39;, &#39;L8TX_180815_01_E07&#39;, &#39;L8TX_180815_01_E08&#39;, &#39;L8TX_180815_01_F07&#39;, &#39;L8TX_180815_01_F08&#39;, &#39;L8TX_180815_01_G07&#39;, &#39;L8TX_180815_01_G08&#39;, &#39;L8TX_180815_01_H07&#39;, &#39;L8TX_180815_01_H08&#39;, &#39;L8TX_180829_01_A09&#39;, &#39;L8TX_180829_01_B09&#39;, &#39;L8TX_180829_01_C09&#39;, &#39;L8TX_180829_01_C10&#39;, &#39;L8TX_180829_01_E09&#39;, &#39;L8TX_180829_01_F09&#39;, &#39;L8TX_180829_01_G09&#39;, &#39;L8TX_190321_01_F03&#39;, &#39;L8TX_190321_01_G03&#39;, &#39;L8TX_190321_01_H03&#39;]
Number of unique anatomical_division_label = 1 [&#39;TH&#39;]
Number of unique cluster_alias = 1082 
Number of unique library_method = 1 [&#39;10Xv2&#39;]
Number of unique region_of_interest_acronym = 1 [&#39;TH&#39;]
Number of unique donor_label = 8 [&#39;Snap25-IRES2-Cre;Ai14-404122&#39;, &#39;Snap25-IRES2-Cre;Ai14-404123&#39;, &#39;Snap25-IRES2-Cre;Ai14-404124&#39;, &#39;Snap25-IRES2-Cre;Ai14-404125&#39;, &#39;Snap25-IRES2-Cre;Ai14-407901&#39;, &#39;Snap25-IRES2-Cre;Ai14-407902&#39;, &#39;Snap25-IRES2-Cre;Ai14-407905&#39;, &#39;Snap25-IRES2-Cre;Ai14-449635&#39;]
Number of unique donor_genotype = 1 [&#39;Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt&#39;]
Number of unique donor_sex = 2 [&#39;F&#39;, &#39;M&#39;]
Number of unique dataset_label = 1 [&#39;WMB-10Xv2&#39;]
Number of unique matrix_label = 1 [&#39;WMB-10Xv2-TH&#39;]
Number of unique x = 131038 
Number of unique y = 130920 
Number of unique neurotransmitter = 7 [&#39;Chol&#39;, &#39;Dopa&#39;, &#39;GABA&#39;, &#39;GABA-Glyc&#39;, &#39;Glut&#39;, &#39;Glut-GABA&#39;, &#39;Hist&#39;]
Number of unique division = 7 [&#39;1 Pallium glutamatergic&#39;, &#39;2 Subpallium GABAergic&#39;, &#39;3 PAL-sAMY-TH-HY-MB-HB neuronal&#39;, &#39;4 CBX-MOB-other neuronal&#39;, &#39;5 Neuroglial&#39;, &#39;6 Vascular&#39;, &#39;7 Immune&#39;]
Number of unique class = 25 [&#39;01 IT-ET Glut&#39;, &#39;03 MOB-DG-IMN&#39;, &#39;04 CGE GABA&#39;, &#39;05 MGE GABA&#39;, &#39;06 CNU GABA&#39;, &#39;07 LSX GABA&#39;, &#39;08 MH-LH Glut&#39;, &#39;09 TH Glut&#39;, &#39;11 HY GABA&#39;, &#39;12 MOB-CR Glut&#39;, &#39;13 CNU-HYa Glut&#39;, &#39;14 CNU-HYa GABA&#39;, &#39;15 HY Glut&#39;, &#39;16 MB Glut&#39;, &#39;17 P Glut&#39;, &#39;19 MY Glut&#39;, &#39;20 MB GABA&#39;, &#39;21 P GABA&#39;, &#39;22 MY GABA&#39;, &#39;23 MB Dopa&#39;, &#39;27 Pineal Glut&#39;, &#39;28 Astro-Epen&#39;, &#39;29 Oligo&#39;, &#39;31 Vascular&#39;, &#39;32 Immune&#39;]
Number of unique subclass = 169 
Number of unique supertype = 402 
Number of unique cluster = 1082 
Number of unique neurotransmitter_color = 8 [&#39;#0a9964&#39;, &#39;#2B93DF&#39;, &#39;#73E785&#39;, &#39;#820e57&#39;, &#39;#FF3358&#39;, &#39;#f9f9f9&#39;, &#39;#fcf04b&#39;, &#39;#ff7621&#39;]
Number of unique division_color = 7 [&#39;#4D58FF&#39;, &#39;#53776C&#39;, &#39;#94AF97&#39;, &#39;#9F3DCC&#39;, &#39;#A6906F&#39;, &#39;#FF2678&#39;, &#39;#fc810f&#39;]
Number of unique class_color = 25 [&#39;#0d0177&#39;, &#39;#16f2f2&#39;, &#39;#1ca58c&#39;, &#39;#267300&#39;, &#39;#4498b5&#39;, &#39;#450099&#39;, &#39;#4c97c9&#39;, &#39;#50e5a7&#39;, &#39;#5485c9&#39;, &#39;#64c2fc&#39;, &#39;#6BB8D6&#39;, &#39;#775cce&#39;, &#39;#854cf7&#39;, &#39;#8CCC00&#39;, &#39;#D78082&#39;, &#39;#DAE64A&#39;, &#39;#E5C865&#39;, &#39;#bf87d4&#39;, &#39;#cdef67&#39;, &#39;#ce3b79&#39;, &#39;#e23d66&#39;, &#39;#ed9863&#39;, &#39;#f2882b&#39;, &#39;#f74294&#39;, &#39;#f954ee&#39;]
Number of unique subclass_color = 169 
Number of unique supertype_color = 402 
Number of unique cluster_color = 1082 
Number of unique region_of_interest_order = 1 [23]
Number of unique region_of_interest_color = 1 [&#39;#FF00EF&#39;]
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="example-use-cases">
<h2>Example use cases<a class="headerlink" href="#example-use-cases" title="Permalink to this headline">#</a></h2>
<p>In this section, we explore two use cases. The first example looks at the expression of nine canonical neurotransmitter transporter genes and the second the expression of gene Tac2.</p>
<p>To support these use cases, we will create a smaller submatrix (all cells and 10 genes) that read it into memory. <em>Note this operation takes around 10 seconds</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ntgenes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Slc17a7&#39;</span><span class="p">,</span><span class="s1">&#39;Slc17a6&#39;</span><span class="p">,</span><span class="s1">&#39;Slc17a8&#39;</span><span class="p">,</span><span class="s1">&#39;Slc32a1&#39;</span><span class="p">,</span><span class="s1">&#39;Slc6a5&#39;</span><span class="p">,</span><span class="s1">&#39;Slc18a3&#39;</span><span class="p">,</span><span class="s1">&#39;Slc6a3&#39;</span><span class="p">,</span><span class="s1">&#39;Slc6a4&#39;</span><span class="p">,</span><span class="s1">&#39;Slc6a2&#39;</span><span class="p">]</span>
<span class="n">exgenes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Tac2&#39;</span><span class="p">]</span>
<span class="n">gnames</span> <span class="o">=</span> <span class="n">ntgenes</span> <span class="o">+</span> <span class="n">exgenes</span>
<span class="n">pred</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="ow">in</span> <span class="n">gnames</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">]</span>
<span class="n">gene_filtered</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="n">gene_filtered</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_identifier</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSMUSG00000037771</th>
      <td>Slc32a1</td>
    </tr>
    <tr>
      <th>ENSMUSG00000070570</th>
      <td>Slc17a7</td>
    </tr>
    <tr>
      <th>ENSMUSG00000039728</th>
      <td>Slc6a5</td>
    </tr>
    <tr>
      <th>ENSMUSG00000030500</th>
      <td>Slc17a6</td>
    </tr>
    <tr>
      <th>ENSMUSG00000055368</th>
      <td>Slc6a2</td>
    </tr>
    <tr>
      <th>ENSMUSG00000019935</th>
      <td>Slc17a8</td>
    </tr>
    <tr>
      <th>ENSMUSG00000025400</th>
      <td>Tac2</td>
    </tr>
    <tr>
      <th>ENSMUSG00000020838</th>
      <td>Slc6a4</td>
    </tr>
    <tr>
      <th>ENSMUSG00000021609</th>
      <td>Slc6a3</td>
    </tr>
    <tr>
      <th>ENSMUSG00000100241</th>
      <td>Slc18a3</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">process_time</span><span class="p">()</span>
<span class="n">asubset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span><span class="n">gene_filtered</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">to_memory</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;time taken: &quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">process_time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">asubset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>time taken:  4.070343479000002
AnnData object with n_obs × n_vars = 131212 × 10
    obs: &#39;cell_barcode&#39;, &#39;library_label&#39;, &#39;anatomical_division_label&#39;
    var: &#39;gene_symbol&#39;
    uns: &#39;normalization&#39;, &#39;parent&#39;, &#39;parent_layer&#39;, &#39;parent_rows&#39;
</pre></div>
</div>
</div>
</div>
<p>We define two helper functions to (1) <em>create_expression_dataframe</em>: create joined gene expression and cell metadata dataframe based for a set of input genes and (2) <em>aggregate_by_metadata</em> which computes the average expression for each term in a given category.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_expression_dataframe</span><span class="p">(</span> <span class="n">ad</span><span class="p">,</span> <span class="n">gf</span> <span class="p">)</span> <span class="p">:</span>
    <span class="n">gdata</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[:,</span><span class="n">gf</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
    <span class="n">gdata</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">gf</span><span class="o">.</span><span class="n">gene_symbol</span>
    <span class="n">joined</span> <span class="o">=</span> <span class="n">cell_filtered</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">gdata</span> <span class="p">)</span>
    <span class="k">return</span> <span class="n">joined</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">aggregate_by_metadata</span><span class="p">(</span> <span class="n">df</span><span class="p">,</span> <span class="n">gnames</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span> <span class="p">)</span> <span class="p">:</span>
    <span class="n">grouped</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">value</span><span class="p">)[</span><span class="n">gnames</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">sort</span> <span class="p">:</span>
        <span class="n">grouped</span> <span class="o">=</span> <span class="n">grouped</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">gnames</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">grouped</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="expression-of-canonical-neurotransmitter-transporter-genes-in-the-thalamus">
<h3>Expression of canonical neurotransmitter transporter genes in the thalamus<a class="headerlink" href="#expression-of-canonical-neurotransmitter-transporter-genes-in-the-thalamus" title="Permalink to this headline">#</a></h3>
<p>During analysis, clusters were assigned neurotransmitter identities based on the expression of of canonical neurotransmitter transporter genes. In this example, we create a dataframe comprising of cells in the ‘WMB-10Xv2-TH’ package and expression of the 9 solute carrier family genes. We then group the cells by the assigned neurotransmitter class and compute the mean expression for each group. We define a helper function that display a dataframe as a heatmap for visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_heatmap</span><span class="p">(</span> <span class="n">df</span><span class="p">,</span> <span class="n">fig_width</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Reds&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="p">:</span>

    <span class="n">arr</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="n">fig_width</span><span class="p">,</span><span class="n">fig_height</span><span class="p">)</span>

    <span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">arr</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">)</span>
    <span class="n">xlabs</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span>
    <span class="n">ylabs</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabs</span><span class="p">)))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span> <span class="n">xlabs</span> <span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ylabs</span><span class="p">)))</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span> <span class="n">ylabs</span> <span class="p">)</span>
    
</pre></div>
</div>
</div>
</div>
<p>Within the thalamus, Slc17a7 is most enriched in cholinergic assigned cells with some expression in glutamatergic types. Genes Slc17a6 is enriched in glutamatergic, Slc32a1 in GABAergic, Slc6a5 in glycinergic, Slc18a3 in cholinergic and Slc6a3 in dopaminergic types, matching the expectation described in the manuscript.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="ow">in</span> <span class="n">ntgenes</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">asubset</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">]</span>
<span class="n">gf</span> <span class="o">=</span> <span class="n">asubset</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>

<span class="n">ntexp</span> <span class="o">=</span> <span class="n">create_expression_dataframe</span><span class="p">(</span> <span class="n">asubset</span><span class="p">,</span> <span class="n">gf</span> <span class="p">)</span>
<span class="n">agg</span> <span class="o">=</span> <span class="n">aggregate_by_metadata</span><span class="p">(</span><span class="n">ntexp</span><span class="p">,</span> <span class="n">gf</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="s1">&#39;neurotransmitter&#39;</span><span class="p">)</span>
<span class="n">plot_heatmap</span><span class="p">(</span><span class="n">agg</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="n">vmax</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ae93661af6fb767aa2d22044546699073f02e24e711895b64fc05bfefb13f28b.png" src="../_images/ae93661af6fb767aa2d22044546699073f02e24e711895b64fc05bfefb13f28b.png" />
</div>
</div>
</div>
<div class="section" id="expression-of-tachykinin-2-tac2-in-the-thalamus">
<h3>Expression of Tachykinin 2 (Tac2) in the thalamus<a class="headerlink" href="#expression-of-tachykinin-2-tac2-in-the-thalamus" title="Permalink to this headline">#</a></h3>
<p>In mice, the tachykinin 2 (Tac2) gene encodes neuropeptide called neurokinin B (NkB). Tac2 is produced by neurons in specific regions of the brain know to be invovled in emotion and social behavior. Based on <a class="reference external" href="https://mouse.brain-map.org/experiment/siv?id=77279001&amp;imageId=77284584&amp;initImage=ish&amp;coordSystem=pixel&amp;x=5384.5&amp;y=3832.5&amp;z=3">ISH data</a> from the Allen Mouse Brain Atlas, Tac 2 is sparsely expressed in the mouse isocortex and densely enriched is specific subcortical regions such the medial habenula (MH), the amygdala and hypothalamus.</p>
<p>In this example, we create a dataframe comprising of cells in the ‘WMB-10Xv2-TH’ package and expression values of Tac2 in those cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gf</span> <span class="o">=</span> <span class="n">asubset</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">asubset</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_symbol</span> <span class="o">==</span> <span class="s1">&#39;Tac2&#39;</span><span class="p">]</span>
<span class="n">tac2</span> <span class="o">=</span> <span class="n">create_expression_dataframe</span><span class="p">(</span> <span class="n">asubset</span><span class="p">,</span> <span class="n">gf</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Grouping cells by neurotransmitter identites and computing the mean expression in each group, we can observed that Tac2 gene is highly enriched in cholinergic cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">agg</span> <span class="o">=</span> <span class="n">aggregate_by_metadata</span><span class="p">(</span><span class="n">tac2</span><span class="p">,</span> <span class="n">gf</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="s1">&#39;neurotransmitter&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plot_heatmap</span><span class="p">(</span><span class="n">agg</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/bdc8b5c451d4c50b6f45143e7e3d9cc5322198bfbfb34b4df04e0a045246d518.png" src="../_images/bdc8b5c451d4c50b6f45143e7e3d9cc5322198bfbfb34b4df04e0a045246d518.png" />
</div>
</div>
<p>Grouping cells by cell types class, shows that Tac2 is enriched in class “08 MH-LH Glut” with cells restricted to the medial (MH) and lateral (LH) habenula and a mixture of glutamatergic and cholinergic types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">agg</span> <span class="o">=</span> <span class="n">aggregate_by_metadata</span><span class="p">(</span><span class="n">tac2</span><span class="p">,</span> <span class="n">gf</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
<span class="n">plot_heatmap</span><span class="p">(</span><span class="n">agg</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/70aaa535e14b21e4082add161da74cc825176aa52fe48e35af33ef8a1f88961f.png" src="../_images/70aaa535e14b21e4082add161da74cc825176aa52fe48e35af33ef8a1f88961f.png" />
</div>
</div>
<p>At the next level, grouping by subclass, shows Tac2 is enriched in sublcass “063 MH Tac2 Glut” with further specificity to the medial habenula and with Tac2 itself as a marker for the subclass.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">agg</span> <span class="o">=</span> <span class="n">aggregate_by_metadata</span><span class="p">(</span><span class="n">tac2</span><span class="p">,</span> <span class="n">gf</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="s1">&#39;subclass&#39;</span><span class="p">,</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
<span class="n">plot_heatmap</span><span class="p">(</span><span class="n">agg</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/bcfdd8147ca8199c3a3976fd7697cd490b652acdcce11b6371b7f4b3e957ece4.png" src="../_images/bcfdd8147ca8199c3a3976fd7697cd490b652acdcce11b6371b7f4b3e957ece4.png" />
</div>
</div>
<p>Next level down, gene Tac2 is enriched in several children supertypes of “063 MH Tac2 Glut” with the exception of “0252 MH Tac2 Glut_5”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">agg</span> <span class="o">=</span> <span class="n">aggregate_by_metadata</span><span class="p">(</span><span class="n">tac2</span><span class="p">,</span> <span class="n">gf</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="s1">&#39;supertype&#39;</span><span class="p">,</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plot_heatmap</span><span class="p">(</span><span class="n">agg</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/d9c11214c7f54fcc6ecd8dde11b31556737e7188b86fb293b81273e7511d5078.png" src="../_images/d9c11214c7f54fcc6ecd8dde11b31556737e7188b86fb293b81273e7511d5078.png" />
</div>
</div>
<p>Close h5ad file and clean up</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="k">del</span> <span class="n">adata</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-overview">Data overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-metadata">Cell metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#umap-spatial-embedding">UMAP spatial embedding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-transcriptomes">Single cell transcriptomes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-expression-matrices">Gene expression matrices</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-use-cases">Example use cases</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-of-canonical-neurotransmitter-transporter-genes-in-the-thalamus">Expression of canonical neurotransmitter transporter genes in the thalamus</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-of-tachykinin-2-tac2-in-the-thalamus">Expression of Tachykinin 2 (Tac2) in the thalamus</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Allen Institute
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>