

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>10x scRNA-seq clustering analysis and annotation (CCN20230504) &#8212; Allen Brain Cell Atlas - Data Access</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/cluster_annotation_tutorial';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Cluster neighborhoods and embeddings" href="cluster_groups_and_embeddings_tutorial.html" />
    <link rel="prev" title="MERFISH CCF mapped coordinates" href="../descriptions/MERFISH-C57BL6J-638850-CCF.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/allen_logo.jpeg" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../_static/allen_logo.jpeg" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Allen Brain Cell Atlas
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="getting_started.html">Getting started</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10Xv2.html">Whole mouse brain 10Xv2 single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10Xv3.html">Whole mouse brain 10Xv3 single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10X.html">Clustering analysis of whole adult mouse brain 10X single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-taxonomy.html">Whole adult mouse brain taxonomy of cell types</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/MERFISH-C57BL6J-638850.html">MERFISH spatial transcriptomics dataset of a single adult mouse brain</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-neighborhoods.html">Cluster groups and neighborhood specific UMAP embeddings</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/Allen-CCF-2020.html">Allen Mouse Common Coordinate Framework (2020 version)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/MERFISH-C57BL6J-638850-CCF.html">MERFISH CCF mapped coordinates</a></li>
</ul>
</li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">10x scRNA-seq clustering analysis and annotation (CCN20230504)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_groups_and_embeddings_tutorial.html">Cluster neighborhoods and embeddings</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_neighborhood_gallery.html">Cell type neighborhood gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="ccf_and_parcellation_annotation_tutorial.html">Allen Mouse Common Coordinate Framework (2020 version)</a></li>
<li class="toctree-l1"><a class="reference internal" href="merfish_ccf_registration_tutorial.html">MERFISH CCF mapped coordinates</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fnotebooks/cluster_annotation_tutorial.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/cluster_annotation_tutorial.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>10x scRNA-seq clustering analysis and annotation (CCN20230504)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-overview">Data Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clusters">Clusters</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cluster-annotation-term-sets">Cluster annotation term sets</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cluster-annotation-terms">Cluster annotation terms</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cluster-to-cluster-annotation-membership">Cluster to cluster annotation membership</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-use-cases">Example use cases</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#aggregating-cluster-and-cells-counts-per-term">Aggregating cluster and cells counts per term</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#neurotransmitter-cluster-and-cell-counts">Neurotransmitter cluster and  cell counts</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#division-cluster-and-cell-counts">Division cluster and cell counts</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#class-level-cluster-and-cell-counts">Class level cluster and cell counts</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-cluster-and-term-distributions">Visualizing cluster and term distributions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-neurotransmitter-clusters-in-each-division">Distribution of neurotransmitter clusters in each division</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-neurotransmitter-clusters-in-each-class">Distribution of neurotransmitter clusters in each class</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-division-clusters-in-each-class">Distribution of division clusters in each class</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-the-mouse-whole-brain-taxonomy">Visualizing the mouse whole brain taxonomy</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <div class="tex2jax_ignore mathjax_ignore section" id="x-scrna-seq-clustering-analysis-and-annotation-ccn20230504">
<h1>10x scRNA-seq clustering analysis and annotation (CCN20230504)<a class="headerlink" href="#x-scrna-seq-clustering-analysis-and-annotation-ccn20230504" title="Permalink to this headline">#</a></h1>
<p>A set of 781 10X scRNA-seq libraries were generated from anatomically defined tissue microdissections in adult mouse brains, resulting in a dataset of ~7.0 million single-cell transcriptomes. An iterative clustering analysis (scrattch.bigcat package) was conducted on ~4.3 milion QC-qualified cells resulting 5,283 clusters. The clusters where then computationally grouped into ~307 subclasses. A detailed analysis of all the subclasses and clusters was conducted. During this process, low quality noisy clusters (usally doublets or mixed debris) were identified and removed resulting a final set of 5196 clusters containing a total ~4.0 million high-quality single-cell transcriptomes.</p>
<p>To organize the complex molecular relationships, a hierarchical representation of cell types was defined with 5 nested levels of classification: division, class, subclass, supertype and clusters. Further neurotransmitter identity was assigned to each cluster based on expression of canonical neurotransmitter transporter genes. See <a class="reference external" href="https://doi.org/10.1101/2023.03.06.531121">Yao et al</a> for further details about the data, methods and results.</p>
<p>The purpose of this notebook is to provide an overview of how cluster and cluster annotation information is represented through example use cases.</p>
<p>You need to be connected to the internet to run this notebook and that you have already downloaded the data via the getting started notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
</pre></div>
</div>
</div>
</div>
<p>The prerequisite for running this notebook is that the data have been downloaded to local directory maintaining the organization from the manifest.json. <strong>Change the download_base variable to where you have downloaded the data in your system.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">download_base</span> <span class="o">=</span> <span class="s1">&#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com&#39;</span>

<span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/releases/20230630/manifest.json&#39;</span>
<span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
    
<span class="n">metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;WMB-taxonomy&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">view_directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> 
                               <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;directory_listing&#39;</span><span class="p">][</span><span class="s1">&#39;WMB-taxonomy&#39;</span><span class="p">][</span><span class="s1">&#39;directories&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">],</span> 
                              <span class="s1">&#39;views&#39;</span><span class="p">)</span>
<span class="n">view_directory</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span> <span class="n">view_directory</span> <span class="p">)</span>
<span class="n">cache_views</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="data-overview">
<h2>Data Overview<a class="headerlink" href="#data-overview" title="Permalink to this headline">#</a></h2>
<div class="section" id="clusters">
<h3>Clusters<a class="headerlink" href="#clusters" title="Permalink to this headline">#</a></h3>
<p>Each of the final set of 5196 cluster is associated with an alias and label. Each row of the dataframe represnts a <em><strong>cluster</strong></em>. Each <em>cluster</em> has a <em>label</em> (human readable string that is unique in the database), <em>cluster alias</em> (in this case a simple integer) and the <em>number of cells</em> that has been grouped into the cluster.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;cluster&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">cluster</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">cluster</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cluster_alias</th>
      <th>number_of_cells</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>731</td>
      <td>CS20230305_0001</td>
    </tr>
    <tr>
      <th>1</th>
      <td>10</td>
      <td>741</td>
      <td>CS20230305_0010</td>
    </tr>
    <tr>
      <th>2</th>
      <td>100</td>
      <td>1041</td>
      <td>CS20230305_0100</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1000</td>
      <td>59</td>
      <td>CS20230305_1000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1001</td>
      <td>96</td>
      <td>CS20230305_1001</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="cluster-annotation-term-sets">
<h3>Cluster annotation term sets<a class="headerlink" href="#cluster-annotation-term-sets" title="Permalink to this headline">#</a></h3>
<p>Each level of classification is represented as a <em><strong>cluster annotation term set</strong></em>. Each term set consists of a set of ordered terms. Each term set has a <em>label</em> (human readable string that is unique in the database), a <em>name</em>, <em>description</em> and <em>order</em> among the term sets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;cluster_annotation_term_set&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">term_set</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">term_set</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>label</th>
      <th>description</th>
      <th>name</th>
      <th>order</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CCN20230504_NEUR</td>
      <td>Clusters are assigned based on the average exp...</td>
      <td>neurotransmitter</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CCN20230504_DIVI</td>
      <td>A very broad definition of cell types. Cluster...</td>
      <td>division</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CCN20230504_CLAS</td>
      <td>The top level of cell type definition in the m...</td>
      <td>class</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CCN20230504_SUBC</td>
      <td>The coarse level of cell type definition in th...</td>
      <td>subclass</td>
      <td>3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CCN20230504_SUPT</td>
      <td>The second finest level of cell type definitio...</td>
      <td>supertype</td>
      <td>4</td>
    </tr>
    <tr>
      <th>5</th>
      <td>CCN20230504_CLUS</td>
      <td>The finest level of cell type definition in th...</td>
      <td>cluster</td>
      <td>5</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can use the pandas iterrows function to print out the description for each term set.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># print out description for each term_set</span>
<span class="k">for</span> <span class="n">tsindex</span><span class="p">,</span> <span class="n">tsrow</span> <span class="ow">in</span> <span class="n">term_set</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span> <span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">:</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tsrow</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tsrow</span><span class="p">[</span><span class="s1">&#39;description&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>neurotransmitter:

Clusters are assigned based on the average expression of the most dominent neurotransmitter transporters. This definition is not part of the taxonomy hierarchy. 

division:

A very broad definition of cell types. Clusters with continuous transition between each other are grouped in the same division. Clusters within the same class mostly belong to the same division, except for some highly distinct clusters that may belong to a specific &quot;CBX-MOB-other neuronal&quot; division.

class:

The top level of cell type definition in the mouse whole brain taxonomy. It is primarily determined by broad brain region and neurotransmitter type. All cells within a subclass belong to the same class. Class provides a broader categorization of cell types.

subclass:

The coarse level of cell type definition in the mouse whole brain taxonomy. All cells within a supertype belong to the same subclass. Subclass groups together related supertypes.

supertype:

The second finest level of cell type definition in the mouse whole brain taxonomy. All cells within a cluster belong to the same supertype. Supertype groups together similar clusters.

cluster:

The finest level of cell type definition in the mouse whole brain taxonomy. Cells within a cluster share similar characteristics and belong to the same supertype.
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="cluster-annotation-terms">
<h3>Cluster annotation terms<a class="headerlink" href="#cluster-annotation-terms" title="Permalink to this headline">#</a></h3>
<p>A <em><strong>cluster annotation term</strong></em> represents a group within a single level of classification. Each term object has a <em>label</em> (human readable string that is unique in the database), a <em>name</em>, which <em>cluster annotation term set</em> it belongs to and a <em>parent</em> or superclass if the term forms a heirarchy, an order within a <em>term set</em> and a <em>color</em> for visualization purposes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;cluster_annotation_term&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">term</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">term</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>label</th>
      <th>name</th>
      <th>cluster_annotation_term_set_label</th>
      <th>parent_term_label</th>
      <th>parent_term_set_label</th>
      <th>term_set_order</th>
      <th>term_order</th>
      <th>cluster_annotation_term_set_name</th>
      <th>color_hex_triplet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CCN20230504_NEUR_Glut</td>
      <td>Glut</td>
      <td>CCN20230504_NEUR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0</td>
      <td>0</td>
      <td>neurotransmitter</td>
      <td>#2B93DF</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CCN20230504_NEUR_GABA</td>
      <td>GABA</td>
      <td>CCN20230504_NEUR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0</td>
      <td>1</td>
      <td>neurotransmitter</td>
      <td>#FF3358</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CCN20230504_NEUR_Glut-GABA</td>
      <td>Glut-GABA</td>
      <td>CCN20230504_NEUR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0</td>
      <td>2</td>
      <td>neurotransmitter</td>
      <td>#0a9964</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CCN20230504_NEUR_Chol</td>
      <td>Chol</td>
      <td>CCN20230504_NEUR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0</td>
      <td>3</td>
      <td>neurotransmitter</td>
      <td>#73E785</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CCN20230504_NEUR_Dopa</td>
      <td>Dopa</td>
      <td>CCN20230504_NEUR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0</td>
      <td>4</td>
      <td>neurotransmitter</td>
      <td>#fcf04b</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can use the groupby and count functions in pandas to tally the number of terms in each term set (classification level).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">term</span><span class="p">[[</span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>label</th>
    </tr>
    <tr>
      <th>cluster_annotation_term_set_name</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>class</th>
      <td>32</td>
    </tr>
    <tr>
      <th>cluster</th>
      <td>5196</td>
    </tr>
    <tr>
      <th>division</th>
      <td>7</td>
    </tr>
    <tr>
      <th>neurotransmitter</th>
      <td>9</td>
    </tr>
    <tr>
      <th>subclass</th>
      <td>307</td>
    </tr>
    <tr>
      <th>supertype</th>
      <td>1115</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>It is important to note that <strong>term names are only unique within a term set and not unique across term sets</strong>. In particular, the name “0001 Car3 Glut_1” is used both as supertype and cluster term. These are distinct entities as supertype “0001 Car3 Glut_1” has three other distinct cluster members. As such when doing aggregation operations, it is important to the use the term <em><strong>label</strong></em> instead of <em><strong>name</strong></em>. For the name “0001 Car3 Glut_1”, the associated label terms are “CCN20230504_SUPT_0001” and “CCN20230504_CLUS_0001” in the supertype and class level respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the groupby function to find the usage count for each term name</span>
<span class="n">term_name_count</span> <span class="o">=</span> <span class="n">term</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;name&#39;</span><span class="p">])[[</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
<span class="n">term_name_count</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;usage_count&#39;</span><span class="p">]</span>
<span class="n">pred</span> <span class="o">=</span> <span class="n">term_name_count</span><span class="p">[</span><span class="s1">&#39;usage_count&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span>
<span class="n">term_name_count</span> <span class="o">=</span> <span class="n">term_name_count</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="n">term_name_count</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>usage_count</th>
    </tr>
    <tr>
      <th>name</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0001 Car3 Glut_1</th>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter the dataframe to terms where there was a name reuse</span>
<span class="n">pred</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="ow">in</span> <span class="n">term_name_count</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">term</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span>
<span class="n">name_clash_term</span> <span class="o">=</span> <span class="n">term</span><span class="p">[</span><span class="n">pred</span><span class="p">][[</span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">]]</span>
<span class="n">name_clash_term</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>label</th>
      <th>name</th>
      <th>cluster_annotation_term_set_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>355</th>
      <td>CCN20230504_SUPT_0001</td>
      <td>0001 Car3 Glut_1</td>
      <td>supertype</td>
    </tr>
    <tr>
      <th>1470</th>
      <td>CCN20230504_CLUS_0001</td>
      <td>0001 Car3 Glut_1</td>
      <td>cluster</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># List all clusters which have supertype &quot;0001 Car3 Glut_1&quot; as parent</span>
<span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">term</span><span class="p">[</span><span class="s1">&#39;parent_term_label&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CCN20230504_SUPT_0001&#39;</span><span class="p">)</span>
<span class="n">term</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>label</th>
      <th>name</th>
      <th>cluster_annotation_term_set_label</th>
      <th>parent_term_label</th>
      <th>parent_term_set_label</th>
      <th>term_set_order</th>
      <th>term_order</th>
      <th>cluster_annotation_term_set_name</th>
      <th>color_hex_triplet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1470</th>
      <td>CCN20230504_CLUS_0001</td>
      <td>0001 Car3 Glut_1</td>
      <td>CCN20230504_CLUS</td>
      <td>CCN20230504_SUPT_0001</td>
      <td>CCN20230504_SUPT</td>
      <td>5</td>
      <td>0</td>
      <td>cluster</td>
      <td>#3DCCB7</td>
    </tr>
    <tr>
      <th>1471</th>
      <td>CCN20230504_CLUS_0002</td>
      <td>0002 Car3 Glut_1</td>
      <td>CCN20230504_CLUS</td>
      <td>CCN20230504_SUPT_0001</td>
      <td>CCN20230504_SUPT</td>
      <td>5</td>
      <td>1</td>
      <td>cluster</td>
      <td>#456C99</td>
    </tr>
    <tr>
      <th>1472</th>
      <td>CCN20230504_CLUS_0003</td>
      <td>0003 Car3 Glut_1</td>
      <td>CCN20230504_CLUS</td>
      <td>CCN20230504_SUPT_0001</td>
      <td>CCN20230504_SUPT</td>
      <td>5</td>
      <td>2</td>
      <td>cluster</td>
      <td>#e54bd6</td>
    </tr>
    <tr>
      <th>1473</th>
      <td>CCN20230504_CLUS_0004</td>
      <td>0004 Car3 Glut_1</td>
      <td>CCN20230504_CLUS</td>
      <td>CCN20230504_SUPT_0001</td>
      <td>CCN20230504_SUPT</td>
      <td>5</td>
      <td>3</td>
      <td>cluster</td>
      <td>#99672E</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="cluster-to-cluster-annotation-membership">
<h3>Cluster to cluster annotation membership<a class="headerlink" href="#cluster-to-cluster-annotation-membership" title="Permalink to this headline">#</a></h3>
<p>The association between a <em><strong>cluster</strong></em> and <em><strong>cluster annotation term</strong></em> is represented as a <em><strong>cluster to cluster annotation membership</strong></em> within the context of one <em><strong>cluster annotation term set</strong></em>. It is expected that a cluster in only associated with one term within a specific term set.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;cluster_to_cluster_annotation_membership&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">membership</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">membership</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cluster_annotation_term_label</th>
      <th>cluster_annotation_term_set_label</th>
      <th>cluster_alias</th>
      <th>cluster_annotation_term_name</th>
      <th>cluster_annotation_term_set_name</th>
      <th>number_of_cells</th>
      <th>color_hex_triplet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CCN20230504_CLUS_0001</td>
      <td>CCN20230504_CLUS</td>
      <td>128</td>
      <td>0001 Car3 Glut_1</td>
      <td>cluster</td>
      <td>4287</td>
      <td>#3DCCB7</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CCN20230504_CLUS_0002</td>
      <td>CCN20230504_CLUS</td>
      <td>129</td>
      <td>0002 Car3 Glut_1</td>
      <td>cluster</td>
      <td>3232</td>
      <td>#456C99</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CCN20230504_CLUS_0003</td>
      <td>CCN20230504_CLUS</td>
      <td>130</td>
      <td>0003 Car3 Glut_1</td>
      <td>cluster</td>
      <td>12285</td>
      <td>#e54bd6</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CCN20230504_CLUS_0004</td>
      <td>CCN20230504_CLUS</td>
      <td>143</td>
      <td>0004 Car3 Glut_1</td>
      <td>cluster</td>
      <td>9408</td>
      <td>#99672E</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CCN20230504_CLUS_0005</td>
      <td>CCN20230504_CLUS</td>
      <td>131</td>
      <td>0005 Car3 Glut_2</td>
      <td>cluster</td>
      <td>1057</td>
      <td>#CC9E5C</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
</div>
<div class="section" id="example-use-cases">
<h2>Example use cases<a class="headerlink" href="#example-use-cases" title="Permalink to this headline">#</a></h2>
<div class="section" id="aggregating-cluster-and-cells-counts-per-term">
<h3>Aggregating cluster and cells counts per term<a class="headerlink" href="#aggregating-cluster-and-cells-counts-per-term" title="Permalink to this headline">#</a></h3>
<p>We can obtain cluster and cell counts per cluster annotation term using the pandas groupby function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Count the number of clusters associated with each cluster annotation term</span>
<span class="n">term_cluster_count</span> <span class="o">=</span> <span class="n">membership</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cluster_annotation_term_label&#39;</span><span class="p">])[[</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
<span class="n">term_cluster_count</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;number_of_clusters&#39;</span><span class="p">]</span>
<span class="n">term_cluster_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_clusters</th>
    </tr>
    <tr>
      <th>cluster_annotation_term_label</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CCN20230504_CLAS_01</th>
      <td>402</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_02</th>
      <td>83</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_03</th>
      <td>121</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_04</th>
      <td>101</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_05</th>
      <td>105</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sum up the number of cells associated with each cluster annotation term</span>
<span class="n">term_cell_count</span> <span class="o">=</span> <span class="n">membership</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cluster_annotation_term_label&#39;</span><span class="p">])[[</span><span class="s1">&#39;number_of_cells&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">term_cell_count</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;number_of_cells&#39;</span><span class="p">]</span>
<span class="n">term_cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>cluster_annotation_term_label</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CCN20230504_CLAS_01</th>
      <td>1098949</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_02</th>
      <td>311327</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_03</th>
      <td>192570</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_04</th>
      <td>139953</td>
    </tr>
    <tr>
      <th>CCN20230504_CLAS_05</th>
      <td>122878</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Join counts with the term dataframe</span>
<span class="n">term_by_label</span> <span class="o">=</span> <span class="n">term</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;label&#39;</span><span class="p">)</span>
<span class="n">term_with_counts</span> <span class="o">=</span> <span class="n">term_by_label</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">term_cluster_count</span> <span class="p">)</span>
<span class="n">term_with_counts</span> <span class="o">=</span> <span class="n">term_with_counts</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">term_cell_count</span> <span class="p">)</span>
<span class="n">term_with_counts</span><span class="p">[[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">,</span><span class="s1">&#39;number_of_clusters&#39;</span><span class="p">,</span><span class="s1">&#39;number_of_cells&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>cluster_annotation_term_set_name</th>
      <th>number_of_clusters</th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CCN20230504_NEUR_Glut</th>
      <td>Glut</td>
      <td>neurotransmitter</td>
      <td>2537</td>
      <td>2062320</td>
    </tr>
    <tr>
      <th>CCN20230504_NEUR_GABA</th>
      <td>GABA</td>
      <td>neurotransmitter</td>
      <td>2008</td>
      <td>840170</td>
    </tr>
    <tr>
      <th>CCN20230504_NEUR_Glut-GABA</th>
      <td>Glut-GABA</td>
      <td>neurotransmitter</td>
      <td>42</td>
      <td>6444</td>
    </tr>
    <tr>
      <th>CCN20230504_NEUR_Chol</th>
      <td>Chol</td>
      <td>neurotransmitter</td>
      <td>34</td>
      <td>5175</td>
    </tr>
    <tr>
      <th>CCN20230504_NEUR_Dopa</th>
      <td>Dopa</td>
      <td>neurotransmitter</td>
      <td>44</td>
      <td>7504</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>For convenience, we can cache this view for later reuse.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="s1">&#39;cluster_annotation_term_with_counts.csv&#39;</span><span class="p">)</span>
    <span class="n">term_with_counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s visualize cluster and cells counts for of the classification levels using bar plots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">bar_plot_by_level_and_type</span><span class="p">(</span> <span class="n">df</span><span class="p">,</span> <span class="n">level</span><span class="p">,</span> <span class="n">fig_width</span> <span class="o">=</span> <span class="mf">8.5</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span> <span class="p">:</span>
    
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="n">fig_width</span><span class="p">,</span> <span class="n">fig_height</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">ctype</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">([</span><span class="s1">&#39;clusters&#39;</span><span class="p">,</span><span class="s1">&#39;cells&#39;</span><span class="p">])</span> <span class="p">:</span>

        <span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">level</span> <span class="p">)</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">pred</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">pred</span><span class="p">][</span><span class="s1">&#39;number_of_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">ctype</span><span class="p">]</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">pred</span><span class="p">][</span><span class="s1">&#39;color_hex_triplet&#39;</span><span class="p">]</span>
        
        <span class="n">ax</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">names</span><span class="p">,</span><span class="n">counts</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Number of </span><span class="si">%s</span><span class="s1"> by </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">ctype</span><span class="p">,</span><span class="n">level</span><span class="p">)),</span>
        
        <span class="k">if</span> <span class="n">idx</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">:</span>
            <span class="n">ax</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([])</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="neurotransmitter-cluster-and-cell-counts">
<h4>Neurotransmitter cluster and  cell counts<a class="headerlink" href="#neurotransmitter-cluster-and-cell-counts" title="Permalink to this headline">#</a></h4>
<p>The majority of clusters and cells are of glutamatergic, GABAergic or GABA-glycinergic neurotransmitter types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bar_plot_by_level_and_type</span><span class="p">(</span><span class="n">term_with_counts</span><span class="p">,</span><span class="s1">&#39;neurotransmitter&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/c58b1868c0ebbe6a4e4f639ec10c65b08d18ad98d134a1a24a5f29fa98737ea0.png" src="../_images/c58b1868c0ebbe6a4e4f639ec10c65b08d18ad98d134a1a24a5f29fa98737ea0.png" />
</div>
</div>
</div>
<div class="section" id="division-cluster-and-cell-counts">
<h4>Division cluster and cell counts<a class="headerlink" href="#division-cluster-and-cell-counts" title="Permalink to this headline">#</a></h4>
<p>Division “3 PAL-sAMY-TH-HY-MB-HB neuronal” contains the largest number of clusters (3871), while division “1 Pallium glutamatergic” contains the largest number of cells (&gt; 1.4 million).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bar_plot_by_level_and_type</span><span class="p">(</span><span class="n">term_with_counts</span><span class="p">,</span><span class="s1">&#39;division&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/4d029c158bdfb1900f393589b9136ecf9320fb90d7f9b1fbba8438849bc80b7c.png" src="../_images/4d029c158bdfb1900f393589b9136ecf9320fb90d7f9b1fbba8438849bc80b7c.png" />
</div>
</div>
</div>
<div class="section" id="class-level-cluster-and-cell-counts">
<h4>Class level cluster and cell counts<a class="headerlink" href="#class-level-cluster-and-cell-counts" title="Permalink to this headline">#</a></h4>
<p>Class “16 MB Glut” contains the largest number of clusters (&gt;600), while class “01 IT-ET Glut” contains the largest number of cells (&gt;1.0 million).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bar_plot_by_level_and_type</span><span class="p">(</span><span class="n">term_with_counts</span><span class="p">,</span><span class="s1">&#39;class&#39;</span><span class="p">,</span><span class="mf">8.5</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/019df8f7f529b2063ebc5c088756cd82a0b055b15f1548804ac7b50dbc3e690e.png" src="../_images/019df8f7f529b2063ebc5c088756cd82a0b055b15f1548804ac7b50dbc3e690e.png" />
</div>
</div>
</div>
</div>
<div class="section" id="visualizing-cluster-and-term-distributions">
<h3>Visualizing cluster and term distributions<a class="headerlink" href="#visualizing-cluster-and-term-distributions" title="Permalink to this headline">#</a></h3>
<p>We can explore the relationship and distribution of clusters between term sets by creating a pivot table using pandas groupby fuunction. Each row of the resulting dataframe represents a cluster, each column represents a term set and the value in the table is the name of the term that has been associated with the cluster for that specific term set.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pivot</span> <span class="o">=</span> <span class="n">membership</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">,</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">])[</span><span class="s1">&#39;cluster_annotation_term_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">unstack</span><span class="p">()</span>
<span class="n">pivot</span> <span class="o">=</span> <span class="n">pivot</span><span class="p">[</span><span class="n">term_set</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span> <span class="c1"># order columns</span>
<span class="n">pivot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>cluster_annotation_term_set_name</th>
      <th>neurotransmitter</th>
      <th>division</th>
      <th>class</th>
      <th>subclass</th>
      <th>supertype</th>
      <th>cluster</th>
    </tr>
    <tr>
      <th>cluster_alias</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0327 L2 IT PPP-APr Glut_3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0081 L2 IT PPP-APr Glut_3</td>
      <td>0328 L2 IT PPP-APr Glut_3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0080 L2 IT PPP-APr Glut_2</td>
      <td>0323 L2 IT PPP-APr Glut_2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0080 L2 IT PPP-APr Glut_2</td>
      <td>0324 L2 IT PPP-APr Glut_2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Glut</td>
      <td>1 Pallium glutamatergic</td>
      <td>01 IT-ET Glut</td>
      <td>018 L2 IT PPP-APr Glut</td>
      <td>0080 L2 IT PPP-APr Glut_2</td>
      <td>0326 L2 IT PPP-APr Glut_2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5406</th>
      <td>Glut</td>
      <td>3 PAL-sAMY-TH-HY-MB-HB neuronal</td>
      <td>19 MY Glut</td>
      <td>180 MB/HB Lhx1 Glut</td>
      <td>0740 MB/HB Lhx1 Glut_1</td>
      <td>3711 MB/HB Lhx1 Glut_1</td>
    </tr>
    <tr>
      <th>5407</th>
      <td>Glut</td>
      <td>3 PAL-sAMY-TH-HY-MB-HB neuronal</td>
      <td>19 MY Glut</td>
      <td>180 MB/HB Lhx1 Glut</td>
      <td>0741 MB/HB Lhx1 Glut_2</td>
      <td>3722 MB/HB Lhx1 Glut_2</td>
    </tr>
    <tr>
      <th>5408</th>
      <td>Glut</td>
      <td>3 PAL-sAMY-TH-HY-MB-HB neuronal</td>
      <td>16 MB Glut</td>
      <td>158 PRC-PAG Tcf7l2 Glut</td>
      <td>0636 PRC-PAG Tcf7l2 Glut_2</td>
      <td>2911 PRC-PAG Tcf7l2 Glut_2</td>
    </tr>
    <tr>
      <th>5409</th>
      <td>Glut</td>
      <td>3 PAL-sAMY-TH-HY-MB-HB neuronal</td>
      <td>16 MB Glut</td>
      <td>158 PRC-PAG Tcf7l2 Glut</td>
      <td>0639 PRC-PAG Tcf7l2 Glut_5</td>
      <td>2921 PRC-PAG Tcf7l2 Glut_5</td>
    </tr>
    <tr>
      <th>5411</th>
      <td>Glut</td>
      <td>3 PAL-sAMY-TH-HY-MB-HB neuronal</td>
      <td>16 MB Glut</td>
      <td>156 APN C1ql4 Glut</td>
      <td>0629 APN C1ql4 Glut_2</td>
      <td>2872 APN C1ql4 Glut_2</td>
    </tr>
  </tbody>
</table>
<p>5196 rows × 6 columns</p>
</div></div></div>
</div>
<p>We can also obtain a cluster annotation color pivot table in the same way.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">color</span> <span class="o">=</span> <span class="n">membership</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">,</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">])[</span><span class="s1">&#39;color_hex_triplet&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">unstack</span><span class="p">()</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;#f9f9f9&#39;</span><span class="p">)</span>
<span class="n">color</span> <span class="o">=</span> <span class="n">color</span><span class="p">[</span><span class="n">term_set</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span> <span class="c1"># order columns</span>
<span class="n">color</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_color&#39;</span> <span class="o">%</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">color</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<span class="n">color</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>neurotransmitter_color</th>
      <th>division_color</th>
      <th>class_color</th>
      <th>subclass_color</th>
      <th>supertype_color</th>
      <th>cluster_color</th>
    </tr>
    <tr>
      <th>cluster_alias</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#26FF8C</td>
    </tr>
    <tr>
      <th>2</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#266DFF</td>
      <td>#280099</td>
    </tr>
    <tr>
      <th>3</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#002BCC</td>
      <td>#5CCCA0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#002BCC</td>
      <td>#CC6D1F</td>
    </tr>
    <tr>
      <th>5</th>
      <td>#2B93DF</td>
      <td>#4D58FF</td>
      <td>#f74294</td>
      <td>#0F6632</td>
      <td>#002BCC</td>
      <td>#FF8E26</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5406</th>
      <td>#2B93DF</td>
      <td>#9F3DCC</td>
      <td>#D78082</td>
      <td>#FF000F</td>
      <td>#FF2689</td>
      <td>#E22600</td>
    </tr>
    <tr>
      <th>5407</th>
      <td>#2B93DF</td>
      <td>#9F3DCC</td>
      <td>#D78082</td>
      <td>#FF000F</td>
      <td>#0F2466</td>
      <td>#5D0099</td>
    </tr>
    <tr>
      <th>5408</th>
      <td>#2B93DF</td>
      <td>#9F3DCC</td>
      <td>#1ca58c</td>
      <td>#30660F</td>
      <td>#660007</td>
      <td>#5CCCA4</td>
    </tr>
    <tr>
      <th>5409</th>
      <td>#2B93DF</td>
      <td>#9F3DCC</td>
      <td>#1ca58c</td>
      <td>#30660F</td>
      <td>#CC1F56</td>
      <td>#65CC5C</td>
    </tr>
    <tr>
      <th>5411</th>
      <td>#2B93DF</td>
      <td>#9F3DCC</td>
      <td>#1ca58c</td>
      <td>#005C99</td>
      <td>#FF4D73</td>
      <td>#894D00</td>
    </tr>
  </tbody>
</table>
<p>5196 rows × 6 columns</p>
</div></div></div>
</div>
<p>For convenience, we can cache this view for later reuse.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    
    <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="s1">&#39;cluster_to_cluster_annotation_membership_pivoted.csv&#39;</span><span class="p">)</span>
    <span class="n">pivot</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
    
    <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="s1">&#39;cluster_to_cluster_annotation_membership_color.csv&#39;</span><span class="p">)</span>
    <span class="n">color</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>For a given pair of term sets A and B, define a function <em>distribution</em> that creates a cluster count table where the rows are terms in term set A, columns are terms in term set B and the table values being the number of clusters that is shared between the terms.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">distribution</span><span class="p">(</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span> <span class="p">)</span> <span class="p">:</span>
    
    <span class="n">AxB</span> <span class="o">=</span> <span class="n">pivot</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">A</span><span class="p">,</span><span class="n">B</span><span class="p">])[[</span><span class="s1">&#39;cluster&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">AxB</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;number_of_clusters&#39;</span><span class="p">]</span>
    <span class="n">AxB</span> <span class="o">=</span> <span class="n">AxB</span><span class="o">.</span><span class="n">unstack</span><span class="p">()</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">B_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">AxB</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span>
    <span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">term</span><span class="p">[</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">B</span><span class="p">)</span>
    <span class="n">term_by_name</span> <span class="o">=</span> <span class="n">term</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">)</span>
    <span class="n">B_colors</span> <span class="o">=</span> <span class="n">term_by_name</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">B_names</span><span class="p">,</span><span class="s1">&#39;color_hex_triplet&#39;</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">AxB</span><span class="p">,</span> <span class="n">B_names</span><span class="p">,</span> <span class="n">B_colors</span>
</pre></div>
</div>
</div>
</div>
<p>Function <em>stacked_bar_distribution</em> takes the results of <em>distribution</em> as input to create distribution stacked bar plot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">stacked_bar_distribution</span><span class="p">(</span> <span class="n">AxB</span><span class="p">,</span> <span class="n">B_names</span><span class="p">,</span> <span class="n">B_colors</span><span class="p">,</span> <span class="n">fig_width</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="mi">6</span> <span class="p">)</span> <span class="p">:</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="n">fig_width</span><span class="p">,</span> <span class="n">fig_height</span><span class="p">)</span>

    <span class="n">bottom</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">AxB</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">AxB</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">AxB</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">AxB</span><span class="p">[</span><span class="n">col</span><span class="p">],</span> <span class="n">left</span><span class="o">=</span><span class="n">bottom</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">col</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">B_colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">bottom</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">AxB</span><span class="p">[</span><span class="n">col</span><span class="p">])</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of </span><span class="si">%s</span><span class="s1"> in each </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">AxB</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">names</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">AxB</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="distribution-of-neurotransmitter-clusters-in-each-division">
<h4>Distribution of neurotransmitter clusters in each division<a class="headerlink" href="#distribution-of-neurotransmitter-clusters-in-each-division" title="Permalink to this headline">#</a></h4>
<p>For the stacked bar plot below, it can observed that most divisions are composed of mixed neurotransmitter cluster types with the exception of divison “1 Pallium glutamatergic” which contains only glutamatergic clusters.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">AxB</span><span class="p">,</span> <span class="n">B_names</span><span class="p">,</span> <span class="n">B_colors</span> <span class="o">=</span> <span class="n">distribution</span><span class="p">(</span> <span class="s1">&#39;division&#39;</span><span class="p">,</span> <span class="s1">&#39;neurotransmitter&#39;</span> <span class="p">)</span>
<span class="n">stacked_bar_distribution</span><span class="p">(</span><span class="n">AxB</span><span class="p">,</span><span class="n">B_names</span><span class="p">,</span><span class="n">B_colors</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/25f436fa4a6eb522514693a2c0246c6e37a7fb3e5e8f47f2321b3ccb90f5e585.png" src="../_images/25f436fa4a6eb522514693a2c0246c6e37a7fb3e5e8f47f2321b3ccb90f5e585.png" />
</div>
</div>
</div>
<div class="section" id="distribution-of-neurotransmitter-clusters-in-each-class">
<h4>Distribution of neurotransmitter clusters in each class<a class="headerlink" href="#distribution-of-neurotransmitter-clusters-in-each-class" title="Permalink to this headline">#</a></h4>
<p>At the next level of classification, around 50% of the classes are have a single neurotransmitter identity and remaining 50% are of mixed types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">AxB</span><span class="p">,</span> <span class="n">B_names</span><span class="p">,</span> <span class="n">B_colors</span> <span class="o">=</span> <span class="n">distribution</span><span class="p">(</span> <span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="s1">&#39;neurotransmitter&#39;</span> <span class="p">)</span>
<span class="n">stacked_bar_distribution</span><span class="p">(</span><span class="n">AxB</span><span class="p">,</span><span class="n">B_names</span><span class="p">,</span><span class="n">B_colors</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/8525cae5e494886aec4c08eaac27e6755f2c6d627a383b80e2fab008eb13a5a1.png" src="../_images/8525cae5e494886aec4c08eaac27e6755f2c6d627a383b80e2fab008eb13a5a1.png" />
</div>
</div>
</div>
<div class="section" id="distribution-of-division-clusters-in-each-class">
<h4>Distribution of division clusters in each class<a class="headerlink" href="#distribution-of-division-clusters-in-each-class" title="Permalink to this headline">#</a></h4>
<p>While clusters in a class mostly come from one division there a couple of exceptions. For example, class “03 MOB-DG-IMN” is composed of both “1 Pallium glutamatergic” and “4 CBX-MOB-other neuronal” division clusters.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">AxB</span><span class="p">,</span> <span class="n">B_names</span><span class="p">,</span> <span class="n">B_colors</span> <span class="o">=</span> <span class="n">distribution</span><span class="p">(</span> <span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="s1">&#39;division&#39;</span> <span class="p">)</span>
<span class="n">stacked_bar_distribution</span><span class="p">(</span><span class="n">AxB</span><span class="p">,</span><span class="n">B_names</span><span class="p">,</span><span class="n">B_colors</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/5b1a30e02362c26edb2dc78589ec5109f4acb041b6de2ac6672dafd54c8544dd.png" src="../_images/5b1a30e02362c26edb2dc78589ec5109f4acb041b6de2ac6672dafd54c8544dd.png" />
</div>
</div>
</div>
</div>
<div class="section" id="visualizing-the-mouse-whole-brain-taxonomy">
<h3>Visualizing the mouse whole brain taxonomy<a class="headerlink" href="#visualizing-the-mouse-whole-brain-taxonomy" title="Permalink to this headline">#</a></h3>
<p>Term sets: class, subclass, supertype and cluster forms a four level mouse whole brain taxonomy. We can visualized the taxonomy as a sunburst diagram that shows the single inheritance hierarchy through a series of rings, that are sliced for each annotation term. Each ring corresponds to a level in the hierarchy. We have ordered the rings so that the class level is the outer most ring so that we can add in labels. Rings are sliced up and divided based on their hierarchical relationship to the parent slice. The angle of each slice is proportional to the number of clusters belonging to the term.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">levels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">,</span><span class="s1">&#39;subclass&#39;</span><span class="p">,</span><span class="s1">&#39;supertype&#39;</span><span class="p">]</span>
<span class="n">df</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">lvl</span> <span class="ow">in</span> <span class="n">levels</span> <span class="p">:</span>
    <span class="n">pred</span> <span class="o">=</span> <span class="n">term_with_counts</span><span class="p">[</span><span class="s1">&#39;cluster_annotation_term_set_name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">lvl</span>
    <span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">]</span> <span class="o">=</span> <span class="n">term_with_counts</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
    <span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;parent_term_label&#39;</span><span class="p">,</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span>
<span class="n">size</span> <span class="o">=</span> <span class="mf">0.15</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">lvl</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">levels</span><span class="p">)</span> <span class="p">:</span>
    
    <span class="k">if</span> <span class="n">lvl</span> <span class="o">==</span> <span class="s1">&#39;class&#39;</span> <span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">][</span><span class="s1">&#39;number_of_clusters&#39;</span><span class="p">],</span> <span class="n">colors</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">][</span><span class="s1">&#39;color_hex_triplet&#39;</span><span class="p">],</span>
               <span class="n">labels</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">rotatelabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labeldistance</span><span class="o">=</span><span class="mf">1.025</span><span class="p">,</span>
               <span class="n">radius</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">wedgeprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="kc">None</span><span class="p">),</span><span class="n">startangle</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span> <span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">][</span><span class="s1">&#39;number_of_clusters&#39;</span><span class="p">],</span> <span class="n">colors</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">lvl</span><span class="p">][</span><span class="s1">&#39;color_hex_triplet&#39;</span><span class="p">],</span>
           <span class="n">radius</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">i</span><span class="o">*</span><span class="n">size</span><span class="p">,</span> <span class="n">wedgeprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="kc">None</span><span class="p">),</span><span class="n">startangle</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ff905725a5ba3a233672bfd8cd96f37880c23a12d9dfec6da916e143cdc2085a.png" src="../_images/ff905725a5ba3a233672bfd8cd96f37880c23a12d9dfec6da916e143cdc2085a.png" />
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../descriptions/MERFISH-C57BL6J-638850-CCF.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">MERFISH CCF mapped coordinates</p>
      </div>
    </a>
    <a class="right-next"
       href="cluster_groups_and_embeddings_tutorial.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Cluster neighborhoods and embeddings</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-overview">Data Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clusters">Clusters</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cluster-annotation-term-sets">Cluster annotation term sets</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cluster-annotation-terms">Cluster annotation terms</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cluster-to-cluster-annotation-membership">Cluster to cluster annotation membership</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-use-cases">Example use cases</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#aggregating-cluster-and-cells-counts-per-term">Aggregating cluster and cells counts per term</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#neurotransmitter-cluster-and-cell-counts">Neurotransmitter cluster and  cell counts</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#division-cluster-and-cell-counts">Division cluster and cell counts</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#class-level-cluster-and-cell-counts">Class level cluster and cell counts</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-cluster-and-term-distributions">Visualizing cluster and term distributions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-neurotransmitter-clusters-in-each-division">Distribution of neurotransmitter clusters in each division</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-neurotransmitter-clusters-in-each-class">Distribution of neurotransmitter clusters in each class</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-division-clusters-in-each-class">Distribution of division clusters in each class</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-the-mouse-whole-brain-taxonomy">Visualizing the mouse whole brain taxonomy</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Allen Institute
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>