

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>MERFISH CCF mapped coordinates &#8212; Allen Brain Cell Atlas - Data Access</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/merfish_ccf_registration_tutorial';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Allen Mouse Common Coordinate Framework (2020 version)" href="ccf_and_parcellation_annotation_tutorial.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/allen_logo.jpeg" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../_static/allen_logo.jpeg" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Allen Brain Cell Atlas
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="getting_started.html">Getting started</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10Xv2.html">Whole mouse brain 10Xv2 single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10Xv3.html">Whole mouse brain 10Xv3 single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-10X.html">Clustering analysis of whole adult mouse brain 10X single cell transcriptomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-taxonomy.html">Whole adult mouse brain taxonomy of cell types</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/MERFISH-C57BL6J-638850.html">MERFISH spatial transcriptomics dataset of a single adult mouse brain</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/WMB-neighborhoods.html">Cluster groups and neighborhood specific UMAP embeddings</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/Allen-CCF-2020.html">Allen Mouse Common Coordinate Framework (2020 version)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../descriptions/MERFISH-C57BL6J-638850-CCF.html">MERFISH CCF mapped coordinates</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cluster_annotation_tutorial.html">10x scRNA-seq clustering analysis and annotation (CCN20230504)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_groups_and_embeddings_tutorial.html">Cluster neighborhoods and embeddings</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_neighborhood_gallery.html">Cell type neighborhood gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="ccf_and_parcellation_annotation_tutorial.html">Allen Mouse Common Coordinate Framework (2020 version)</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">MERFISH CCF mapped coordinates</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fnotebooks/merfish_ccf_registration_tutorial.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/merfish_ccf_registration_tutorial.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>MERFISH CCF mapped coordinates</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-in-section-reconstructed-and-ccf-coordinates-for-all-cells">Read in section, reconstructed and CCF coordinates for all cells</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-cells-in-section-reconstruced-and-ccf-coordinates">Visualize cells in section, reconstruced and CCF coordinates</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#associate-mapped-parcellation-and-parcellation-terms-to-the-merfish-data">Associate mapped parcellation and parcellation terms to the MERFISH data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-cell-parcellation-assignment-at-different-anatomical-levels">Visualize cell parcellation assignment at different anatomical levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#aggregate-cells-by-parcellation-assignment-at-different-anatomical-levels">Aggregate cells by parcellation assignment at different anatomical levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#resampled-ccf-template-and-parcellations">Resampled CCF template and parcellations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-cells-with-parcellation-boundary-overlay">Visualize cells with parcellation boundary overlay</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <div class="tex2jax_ignore mathjax_ignore section" id="merfish-ccf-mapped-coordinates">
<h1>MERFISH CCF mapped coordinates<a class="headerlink" href="#merfish-ccf-mapped-coordinates" title="Permalink to this headline">#</a></h1>
<p>Registration to the Allen CCFv3 was performed at 10 micron in-plane resolution. For each section, an anatomical label reference was created using class/subclass level cell type assignments which are then used to match to corresponding CCF parcellations. The midline was manually determined for each section to rotate the section upright and center in in the middle (<strong>section coordinates</strong>). This set of rectified images were stacked in sequential order to create an inital configuration for registration.</p>
<p>The ANTS registration framework was use to establish a 2.5D deformable spatial mapping between the MERFISH data and <strong>CCF coordinates</strong> via three major steps: (1) A 3D global affine (12 dof) mapping was performed to align the CCF into the MERFISH space (<strong>resampled CCF</strong>); (2) 2D affine (6 dof) registration of each MERFISH section to match the target resampled CCF section and finally, (3) a 2D multi-scale, symmetric diffeomorphic registration of each section (<strong>reconstructed coordinates</strong>) to finer level match to the target CCF section. Global and section-wise mappings from each of these registration steps were preserved and concatenated (with appropriate inversions) to allow point-to-point mapping between the original MERFISH coordinate space and the CCF space. See <a class="reference external" href="https://doi.org/10.1101/2023.03.06.531121">Yao et al</a> for further details.</p>
<p>The purpose of this notebook is to provide an overview of how the mapped coordinate information is represented through example use cases.</p>
<p>You need to be connected to the internet to run this notebook and that you have already downloaded the data via the getting started notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">import</span> <span class="nn">SimpleITK</span> <span class="k">as</span> <span class="nn">sitk</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">download_base</span> <span class="o">=</span> <span class="s1">&#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com&#39;</span>

<span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/releases/20230630/manifest.json&#39;</span>
<span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The prerequisite for running this notebook is that the data have been downloaded to local directory maintaining the organization from the manifest.json. <strong>Change the download_base variable to where you have downloaded the data in your system.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">view_directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> 
                               <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;directory_listing&#39;</span><span class="p">][</span><span class="s1">&#39;MERFISH-C57BL6J-638850-CCF&#39;</span><span class="p">][</span><span class="s1">&#39;directories&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">],</span> 
                              <span class="s1">&#39;views&#39;</span><span class="p">)</span>
<span class="n">view_directory</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span> <span class="n">view_directory</span> <span class="p">)</span>
<span class="n">cache_views</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="read-in-section-reconstructed-and-ccf-coordinates-for-all-cells">
<h2>Read in section, reconstructed and CCF coordinates for all cells<a class="headerlink" href="#read-in-section-reconstructed-and-ccf-coordinates-for-all-cells" title="Permalink to this headline">#</a></h2>
<p>Read in the expanded cell metadata table we created in part 1 of the MERFISH tutorial. To keep track of the three type of coordinates we will rename the <strong>section coordinates</strong> (x,y,z) in the dataframe to (x_section, y_section, z_section)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;MERFISH-C57BL6J-638850&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;cell_metadata_with_cluster_annotation&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">cell</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;cell_label&quot;</span><span class="p">:</span><span class="nb">str</span><span class="p">})</span>
<span class="n">cell</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;x&#39;</span><span class="p">:</span><span class="s1">&#39;x_section&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;y_section&#39;</span><span class="p">,</span><span class="s1">&#39;z&#39;</span><span class="p">:</span><span class="s1">&#39;z_section&#39;</span><span class="p">},</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cell</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cell_label&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cell</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>brain_section_label</th>
      <th>cluster_alias</th>
      <th>average_correlation_score</th>
      <th>matrix_label</th>
      <th>donor_label</th>
      <th>low_quality_mapping</th>
      <th>donor_genotype</th>
      <th>donor_sex</th>
      <th>x_section</th>
      <th>y_section</th>
      <th>...</th>
      <th>class</th>
      <th>subclass</th>
      <th>supertype</th>
      <th>cluster</th>
      <th>neurotransmitter_color</th>
      <th>division_color</th>
      <th>class_color</th>
      <th>subclass_color</th>
      <th>supertype_color</th>
      <th>cluster_color</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1015221640100570419</th>
      <td>C57BL6J-638850.01</td>
      <td>4090</td>
      <td>0.582276</td>
      <td>C57BL6J-638850.01</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>5.171280</td>
      <td>6.489884</td>
      <td>...</td>
      <td>22 MY GABA</td>
      <td>282 MY Calcb Chol</td>
      <td>0999 MY Calcb Chol</td>
      <td>5100 MY Calcb Chol</td>
      <td>#73E785</td>
      <td>#9F3DCC</td>
      <td>#E5C865</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
    </tr>
    <tr>
      <th>1015221640100590598</th>
      <td>C57BL6J-638850.01</td>
      <td>4090</td>
      <td>0.410491</td>
      <td>C57BL6J-638850.01</td>
      <td>C57BL6J-638850</td>
      <td>True</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>22 MY GABA</td>
      <td>282 MY Calcb Chol</td>
      <td>0999 MY Calcb Chol</td>
      <td>5100 MY Calcb Chol</td>
      <td>#73E785</td>
      <td>#9F3DCC</td>
      <td>#E5C865</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
    </tr>
    <tr>
      <th>1015221640100820600</th>
      <td>C57BL6J-638850.01</td>
      <td>4090</td>
      <td>0.568130</td>
      <td>C57BL6J-638850.01</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>6.547826</td>
      <td>5.712030</td>
      <td>...</td>
      <td>22 MY GABA</td>
      <td>282 MY Calcb Chol</td>
      <td>0999 MY Calcb Chol</td>
      <td>5100 MY Calcb Chol</td>
      <td>#73E785</td>
      <td>#9F3DCC</td>
      <td>#E5C865</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
    </tr>
    <tr>
      <th>1015221640100580476</th>
      <td>C57BL6J-638850.01</td>
      <td>4090</td>
      <td>0.417244</td>
      <td>C57BL6J-638850.01</td>
      <td>C57BL6J-638850</td>
      <td>True</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>22 MY GABA</td>
      <td>282 MY Calcb Chol</td>
      <td>0999 MY Calcb Chol</td>
      <td>5100 MY Calcb Chol</td>
      <td>#73E785</td>
      <td>#9F3DCC</td>
      <td>#E5C865</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
    </tr>
    <tr>
      <th>1015221640100580189</th>
      <td>C57BL6J-638850.01</td>
      <td>4090</td>
      <td>0.453351</td>
      <td>C57BL6J-638850.01</td>
      <td>C57BL6J-638850</td>
      <td>True</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>22 MY GABA</td>
      <td>282 MY Calcb Chol</td>
      <td>0999 MY Calcb Chol</td>
      <td>5100 MY Calcb Chol</td>
      <td>#73E785</td>
      <td>#9F3DCC</td>
      <td>#E5C865</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 23 columns</p>
</div></div></div>
</div>
<p>Next we read in the <strong>reconstructed coordinates</strong> and mapped parcellation from the metadata directory and join it to the cell dataframe</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;MERFISH-C57BL6J-638850-CCF&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;reconstructed_coordinates&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">reconstructed_coords</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;cell_label&quot;</span><span class="p">:</span><span class="nb">str</span><span class="p">})</span>
<span class="n">reconstructed_coords</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;x&#39;</span><span class="p">:</span><span class="s1">&#39;x_reconstructed&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;y_reconstructed&#39;</span><span class="p">,</span><span class="s1">&#39;z&#39;</span><span class="p">:</span><span class="s1">&#39;z_reconstructed&#39;</span><span class="p">},</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">reconstructed_coords</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cell_label&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">reconstructed_coords</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x_reconstructed</th>
      <th>y_reconstructed</th>
      <th>z_reconstructed</th>
      <th>parcellation_index</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1019171911101460569</th>
      <td>7.143894</td>
      <td>7.890964</td>
      <td>0.8</td>
      <td>945</td>
    </tr>
    <tr>
      <th>1019171911101550321</th>
      <td>4.188673</td>
      <td>7.962972</td>
      <td>0.8</td>
      <td>945</td>
    </tr>
    <tr>
      <th>1019171911100841066</th>
      <td>6.859447</td>
      <td>5.908534</td>
      <td>0.8</td>
      <td>893</td>
    </tr>
    <tr>
      <th>1019171911101400425</th>
      <td>3.952014</td>
      <td>7.564086</td>
      <td>0.8</td>
      <td>842</td>
    </tr>
    <tr>
      <th>1019171911101380264</th>
      <td>2.803546</td>
      <td>7.221688</td>
      <td>0.8</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_joined</span> <span class="o">=</span> <span class="n">cell</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reconstructed_coords</span><span class="p">,</span><span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>
<span class="n">cell_joined</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>brain_section_label</th>
      <th>cluster_alias</th>
      <th>average_correlation_score</th>
      <th>matrix_label</th>
      <th>donor_label</th>
      <th>low_quality_mapping</th>
      <th>donor_genotype</th>
      <th>donor_sex</th>
      <th>x_section</th>
      <th>y_section</th>
      <th>...</th>
      <th>neurotransmitter_color</th>
      <th>division_color</th>
      <th>class_color</th>
      <th>subclass_color</th>
      <th>supertype_color</th>
      <th>cluster_color</th>
      <th>x_reconstructed</th>
      <th>y_reconstructed</th>
      <th>z_reconstructed</th>
      <th>parcellation_index</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1019171911101460569</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.518491</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>7.049393</td>
      <td>7.672834</td>
      <td>...</td>
      <td>#2B93DF</td>
      <td>#fc810f</td>
      <td>#bf87d4</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>7.143894</td>
      <td>7.890964</td>
      <td>0.8</td>
      <td>945</td>
    </tr>
    <tr>
      <th>1019171911101550321</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.579740</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>4.450081</td>
      <td>7.630694</td>
      <td>...</td>
      <td>#2B93DF</td>
      <td>#fc810f</td>
      <td>#bf87d4</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>4.188673</td>
      <td>7.962972</td>
      <td>0.8</td>
      <td>945</td>
    </tr>
    <tr>
      <th>1019171911100841066</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.581533</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>6.732956</td>
      <td>5.669881</td>
      <td>...</td>
      <td>#2B93DF</td>
      <td>#fc810f</td>
      <td>#bf87d4</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>6.859447</td>
      <td>5.908534</td>
      <td>0.8</td>
      <td>893</td>
    </tr>
    <tr>
      <th>1019171911101400425</th>
      <td>C57BL6J-638850.05</td>
      <td>4090</td>
      <td>0.704045</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>4.238757</td>
      <td>7.225909</td>
      <td>...</td>
      <td>#73E785</td>
      <td>#9F3DCC</td>
      <td>#E5C865</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
      <td>3.952014</td>
      <td>7.564086</td>
      <td>0.8</td>
      <td>842</td>
    </tr>
    <tr>
      <th>1019171911101380264</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.602300</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>3.226935</td>
      <td>6.847290</td>
      <td>...</td>
      <td>#2B93DF</td>
      <td>#fc810f</td>
      <td>#bf87d4</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>2.803546</td>
      <td>7.221688</td>
      <td>0.8</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27 columns</p>
</div></div></div>
</div>
<p>We repeat the process for the cell <strong>CCF coordinates</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;MERFISH-C57BL6J-638850-CCF&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;ccf_coordinates&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">ccf_coords</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;cell_label&quot;</span><span class="p">:</span><span class="nb">str</span><span class="p">})</span>
<span class="n">ccf_coords</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;x&#39;</span><span class="p">:</span><span class="s1">&#39;x_ccf&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;y_ccf&#39;</span><span class="p">,</span><span class="s1">&#39;z&#39;</span><span class="p">:</span><span class="s1">&#39;z_ccf&#39;</span><span class="p">},</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ccf_coords</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;parcellation_index&#39;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ccf_coords</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cell_label&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ccf_coords</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x_ccf</th>
      <th>y_ccf</th>
      <th>z_ccf</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1019171911101460569</th>
      <td>12.282330</td>
      <td>6.987808</td>
      <td>7.385773</td>
    </tr>
    <tr>
      <th>1019171911101550321</th>
      <td>12.192214</td>
      <td>7.002155</td>
      <td>4.366855</td>
    </tr>
    <tr>
      <th>1019171911100841066</th>
      <td>12.500341</td>
      <td>4.750392</td>
      <td>7.074634</td>
    </tr>
    <tr>
      <th>1019171911101400425</th>
      <td>12.231647</td>
      <td>6.544816</td>
      <td>4.128568</td>
    </tr>
    <tr>
      <th>1019171911101380264</th>
      <td>12.238502</td>
      <td>6.135836</td>
      <td>2.948194</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_joined</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ccf_coords</span><span class="p">,</span><span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>
<span class="n">cell_joined</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>brain_section_label</th>
      <th>cluster_alias</th>
      <th>average_correlation_score</th>
      <th>matrix_label</th>
      <th>donor_label</th>
      <th>low_quality_mapping</th>
      <th>donor_genotype</th>
      <th>donor_sex</th>
      <th>x_section</th>
      <th>y_section</th>
      <th>...</th>
      <th>subclass_color</th>
      <th>supertype_color</th>
      <th>cluster_color</th>
      <th>x_reconstructed</th>
      <th>y_reconstructed</th>
      <th>z_reconstructed</th>
      <th>parcellation_index</th>
      <th>x_ccf</th>
      <th>y_ccf</th>
      <th>z_ccf</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1019171911101460569</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.518491</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>7.049393</td>
      <td>7.672834</td>
      <td>...</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>7.143894</td>
      <td>7.890964</td>
      <td>0.8</td>
      <td>945</td>
      <td>12.282330</td>
      <td>6.987808</td>
      <td>7.385773</td>
    </tr>
    <tr>
      <th>1019171911101550321</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.579740</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>4.450081</td>
      <td>7.630694</td>
      <td>...</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>4.188673</td>
      <td>7.962972</td>
      <td>0.8</td>
      <td>945</td>
      <td>12.192214</td>
      <td>7.002155</td>
      <td>4.366855</td>
    </tr>
    <tr>
      <th>1019171911100841066</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.581533</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>6.732956</td>
      <td>5.669881</td>
      <td>...</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>6.859447</td>
      <td>5.908534</td>
      <td>0.8</td>
      <td>893</td>
      <td>12.500341</td>
      <td>4.750392</td>
      <td>7.074634</td>
    </tr>
    <tr>
      <th>1019171911101400425</th>
      <td>C57BL6J-638850.05</td>
      <td>4090</td>
      <td>0.704045</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>4.238757</td>
      <td>7.225909</td>
      <td>...</td>
      <td>#FF4D81</td>
      <td>#2E5499</td>
      <td>#91FF4D</td>
      <td>3.952014</td>
      <td>7.564086</td>
      <td>0.8</td>
      <td>842</td>
      <td>12.231647</td>
      <td>6.544816</td>
      <td>4.128568</td>
    </tr>
    <tr>
      <th>1019171911101380264</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.602300</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>3.226935</td>
      <td>6.847290</td>
      <td>...</td>
      <td>#9B7ACC</td>
      <td>#990041</td>
      <td>#4E9900</td>
      <td>2.803546</td>
      <td>7.221688</td>
      <td>0.8</td>
      <td>0</td>
      <td>12.238502</td>
      <td>6.135836</td>
      <td>2.948194</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 30 columns</p>
</div></div></div>
</div>
</div>
<div class="section" id="visualize-cells-in-section-reconstruced-and-ccf-coordinates">
<h2>Visualize cells in section, reconstruced and CCF coordinates<a class="headerlink" href="#visualize-cells-in-section-reconstruced-and-ccf-coordinates" title="Permalink to this headline">#</a></h2>
<p>We define a helper function to visualize cells in a section colorized by either cluster or annotation term and optionally with a boundary overlay</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_section</span><span class="p">(</span> <span class="n">xx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">yy</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">val</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pcmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
                 <span class="n">overlay</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bcmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Greys_r</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
                 <span class="n">fig_width</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="mi">6</span> <span class="p">)</span> <span class="p">:</span>
    
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="n">fig_width</span><span class="p">,</span> <span class="n">fig_height</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">xx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">yy</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">pcmap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">xx</span><span class="p">,</span><span class="n">yy</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">val</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">pcmap</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">xx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">yy</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">cc</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">xx</span><span class="p">,</span><span class="n">yy</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">cc</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>   
        
    <span class="k">if</span> <span class="n">overlay</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">extent</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">bcmap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">overlay</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">bcmap</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="n">extent</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span><span class="n">zorder</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">11</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
    
    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span>
</pre></div>
</div>
</div>
</div>
<p>For example section “C57BL6J-638850.40”, we plot cells by <strong>section, reconstructed and ccf coordinates</strong> colorized by neurotransmitter type assignment. Note that cells from a single brain section remains together at the same z (posterior-to-anterior) value in section and reconstructed coordinates. Since the brain sections are at an angle compared to the CCF, the cells are spread across multiple x (anterior-to-posterior) CCF coordinates.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">brain_section</span> <span class="o">=</span> <span class="s1">&#39;C57BL6J-638850.40&#39;</span>
<span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_joined</span><span class="p">[</span><span class="s1">&#39;brain_section_label&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">brain_section</span> <span class="p">)</span>
<span class="n">section</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">section</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>104679
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;z_section values:&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;z_section&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;z_reconstructed values:&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;z_reconstructed&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;x_ccf values:&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;x_ccf&#39;</span><span class="p">])</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>z_section values: [7.2]
z_reconstructed values: [7.2]
x_ccf values: [6.32598314 6.32656693 6.3266852  ... 7.16574148 7.16579943 7.16598123]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">xx</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;x_section&#39;</span><span class="p">],</span> <span class="n">yy</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;y_section&#39;</span><span class="p">],</span> 
                               <span class="n">cc</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;neurotransmitter_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Neuortransmitter - Section Coordinates&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ba03b9e7a82d7cb4caa8811ba85d6853c099740cabe2d4aa83d106516ac155eb.png" src="../_images/ba03b9e7a82d7cb4caa8811ba85d6853c099740cabe2d4aa83d106516ac155eb.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">xx</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;x_reconstructed&#39;</span><span class="p">],</span> <span class="n">yy</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;y_reconstructed&#39;</span><span class="p">],</span> 
                               <span class="n">cc</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;neurotransmitter_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Neuortransmitter - Reconstructed Coordinates&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ea414fb2a76c549d2f244bcbbf4123d448c3e2bd9af1fd4b48f5dc6189107dd8.png" src="../_images/ea414fb2a76c549d2f244bcbbf4123d448c3e2bd9af1fd4b48f5dc6189107dd8.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">xx</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;z_ccf&#39;</span><span class="p">],</span> <span class="n">yy</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;y_ccf&#39;</span><span class="p">],</span> 
                               <span class="n">cc</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;neurotransmitter_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Neuortransmitter - CCF Coordinates&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/10060470f8365720a639af2dad25067017cbf217edba54b7605186c9d33ba022.png" src="../_images/10060470f8365720a639af2dad25067017cbf217edba54b7605186c9d33ba022.png" />
</div>
</div>
</div>
<div class="section" id="associate-mapped-parcellation-and-parcellation-terms-to-the-merfish-data">
<h2>Associate mapped parcellation and parcellation terms to the MERFISH data<a class="headerlink" href="#associate-mapped-parcellation-and-parcellation-terms-to-the-merfish-data" title="Permalink to this headline">#</a></h2>
<p>Read in the pivoted parcellation annotation information we create in the Allen CCFv3 tutorial and join with the cell metadata dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;Allen-CCF-2020&#39;</span><span class="p">][</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;parcellation_to_parcellation_term_membership_acronym&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">parcellation_annotation</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">parcellation_annotation</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;parcellation_index&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">parcellation_annotation</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;parcellation_</span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="n">parcellation_annotation</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<span class="n">parcellation_annotation</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>parcellation_organ</th>
      <th>parcellation_category</th>
      <th>parcellation_division</th>
      <th>parcellation_structure</th>
      <th>parcellation_substructure</th>
    </tr>
    <tr>
      <th>parcellation_index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>unassigned</td>
    </tr>
    <tr>
      <th>1</th>
      <td>brain</td>
      <td>grey</td>
      <td>HY</td>
      <td>TMv</td>
      <td>TMv</td>
    </tr>
    <tr>
      <th>2</th>
      <td>brain</td>
      <td>grey</td>
      <td>Isocortex</td>
      <td>SSp-m</td>
      <td>SSp-m6b</td>
    </tr>
    <tr>
      <th>5</th>
      <td>brain</td>
      <td>fiber tracts</td>
      <td>lfbs</td>
      <td>cst</td>
      <td>int</td>
    </tr>
    <tr>
      <th>6</th>
      <td>brain</td>
      <td>grey</td>
      <td>P</td>
      <td>PSV</td>
      <td>PSV</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rpath</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;parcellation_to_parcellation_term_membership_color&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;csv&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">parcellation_color</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">parcellation_color</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;parcellation_index&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">parcellation_color</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;parcellation_</span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="n">parcellation_color</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<span class="n">parcellation_color</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>parcellation_organ_color</th>
      <th>parcellation_category_color</th>
      <th>parcellation_division_color</th>
      <th>parcellation_structure_color</th>
      <th>parcellation_substructure_color</th>
    </tr>
    <tr>
      <th>parcellation_index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>#000000</td>
      <td>#000000</td>
      <td>#000000</td>
      <td>#000000</td>
      <td>#000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#E64438</td>
      <td>#FF4C3E</td>
      <td>#FF4C3E</td>
    </tr>
    <tr>
      <th>2</th>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#70FF71</td>
      <td>#188064</td>
      <td>#188064</td>
    </tr>
    <tr>
      <th>5</th>
      <td>#FFFFFF</td>
      <td>#CCCCCC</td>
      <td>#CCCCCC</td>
      <td>#CCCCCC</td>
      <td>#CCCCCC</td>
    </tr>
    <tr>
      <th>6</th>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#FF9B88</td>
      <td>#FFAE6F</td>
      <td>#FFAE6F</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_joined</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">parcellation_annotation</span><span class="p">,</span><span class="n">on</span><span class="o">=</span><span class="s1">&#39;parcellation_index&#39;</span><span class="p">)</span>
<span class="n">cell_joined</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">parcellation_color</span><span class="p">,</span><span class="n">on</span><span class="o">=</span><span class="s1">&#39;parcellation_index&#39;</span><span class="p">)</span>
<span class="n">cell_joined</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>brain_section_label</th>
      <th>cluster_alias</th>
      <th>average_correlation_score</th>
      <th>matrix_label</th>
      <th>donor_label</th>
      <th>low_quality_mapping</th>
      <th>donor_genotype</th>
      <th>donor_sex</th>
      <th>x_section</th>
      <th>y_section</th>
      <th>...</th>
      <th>parcellation_organ</th>
      <th>parcellation_category</th>
      <th>parcellation_division</th>
      <th>parcellation_structure</th>
      <th>parcellation_substructure</th>
      <th>parcellation_organ_color</th>
      <th>parcellation_category_color</th>
      <th>parcellation_division_color</th>
      <th>parcellation_structure_color</th>
      <th>parcellation_substructure_color</th>
    </tr>
    <tr>
      <th>cell_label</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1019171911101460569</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.518491</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>7.049393</td>
      <td>7.672834</td>
      <td>...</td>
      <td>brain</td>
      <td>grey</td>
      <td>MY</td>
      <td>LRN</td>
      <td>LRNm</td>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#FF9BCD</td>
      <td>#FFB3D9</td>
      <td>#FFB3D9</td>
    </tr>
    <tr>
      <th>1019171911101550321</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.579740</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>4.450081</td>
      <td>7.630694</td>
      <td>...</td>
      <td>brain</td>
      <td>grey</td>
      <td>MY</td>
      <td>LRN</td>
      <td>LRNm</td>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#FF9BCD</td>
      <td>#FFB3D9</td>
      <td>#FFB3D9</td>
    </tr>
    <tr>
      <th>1019171911100841066</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.581533</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>6.732956</td>
      <td>5.669881</td>
      <td>...</td>
      <td>brain</td>
      <td>grey</td>
      <td>MY</td>
      <td>ECU</td>
      <td>ECU</td>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#FF9BCD</td>
      <td>#FFA5D2</td>
      <td>#FFA5D2</td>
    </tr>
    <tr>
      <th>1019171911101400425</th>
      <td>C57BL6J-638850.05</td>
      <td>4090</td>
      <td>0.704045</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>4.238757</td>
      <td>7.225909</td>
      <td>...</td>
      <td>brain</td>
      <td>grey</td>
      <td>MY</td>
      <td>PARN</td>
      <td>PARN</td>
      <td>#FFFFFF</td>
      <td>#BFDAE3</td>
      <td>#FF9BCD</td>
      <td>#FFB3D9</td>
      <td>#FFB3D9</td>
    </tr>
    <tr>
      <th>1019171911101380264</th>
      <td>C57BL6J-638850.05</td>
      <td>4218</td>
      <td>0.602300</td>
      <td>C57BL6J-638850.05</td>
      <td>C57BL6J-638850</td>
      <td>False</td>
      <td>wt/wt</td>
      <td>M</td>
      <td>3.226935</td>
      <td>6.847290</td>
      <td>...</td>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>unassigned</td>
      <td>#000000</td>
      <td>#000000</td>
      <td>#000000</td>
      <td>#000000</td>
      <td>#000000</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 40 columns</p>
</div></div></div>
</div>
<p>For convenience, we can cache this view for later reuse.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">cache_views</span> <span class="p">:</span>
    <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">view_directory</span><span class="p">,</span> <span class="s1">&#39;cell_metadata_with_parcellation_annotation.csv&#39;</span><span class="p">)</span>
    <span class="n">cell_joined</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="visualize-cell-parcellation-assignment-at-different-anatomical-levels">
<h2>Visualize cell parcellation assignment at different anatomical levels<a class="headerlink" href="#visualize-cell-parcellation-assignment-at-different-anatomical-levels" title="Permalink to this headline">#</a></h2>
<p>For example section “C57BL6J-638850.40”, we plot cells reconstructed coordinates colorized by parcellation assignment at the division and structure anatomical levels</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_joined</span><span class="p">[</span><span class="s1">&#39;brain_section_label&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">brain_section</span><span class="p">)</span>
<span class="n">section</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">section</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>104679
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">xx</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;x_reconstructed&#39;</span><span class="p">],</span> <span class="n">yy</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;y_reconstructed&#39;</span><span class="p">],</span> 
                               <span class="n">cc</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;parcellation_division_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Division - Reconstructed Coordinates&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/dac4e6553f5ba721fece30da50b3ac25308c04fe5a1556da4f73471dd281c689.png" src="../_images/dac4e6553f5ba721fece30da50b3ac25308c04fe5a1556da4f73471dd281c689.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">xx</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;x_reconstructed&#39;</span><span class="p">],</span> <span class="n">yy</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;y_reconstructed&#39;</span><span class="p">],</span> 
                               <span class="n">cc</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;parcellation_structure_color&#39;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Structure - Reconstructed Coordinates&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a633c7e38d76274c47e60b045ba8005da06e31a0da793a5fa1492aff18921d99.png" src="../_images/a633c7e38d76274c47e60b045ba8005da06e31a0da793a5fa1492aff18921d99.png" />
</div>
</div>
</div>
<div class="section" id="aggregate-cells-by-parcellation-assignment-at-different-anatomical-levels">
<h2>Aggregate cells by parcellation assignment at different anatomical levels<a class="headerlink" href="#aggregate-cells-by-parcellation-assignment-at-different-anatomical-levels" title="Permalink to this headline">#</a></h2>
<p>Define a helper function to count the number of cells for each term and apply it to the division and structure anatomical levels</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">parcellation_cell_count</span> <span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">level</span><span class="p">)</span> <span class="p">:</span>
    <span class="n">cell_count</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">level</span><span class="p">)[[</span><span class="s1">&#39;cluster_alias&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">cell_count</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;number_of_cells&#39;</span><span class="p">]</span>
    <span class="n">cell_count</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;number_of_cells&#39;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">cell_count</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_count</span> <span class="o">=</span> <span class="n">parcellation_cell_count</span><span class="p">(</span><span class="n">cell_joined</span><span class="p">,</span><span class="s1">&#39;parcellation_division&#39;</span><span class="p">)</span>
<span class="n">cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>parcellation_division</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Isocortex</th>
      <td>929222</td>
    </tr>
    <tr>
      <th>STR</th>
      <td>400056</td>
    </tr>
    <tr>
      <th>CB</th>
      <td>383781</td>
    </tr>
    <tr>
      <th>HPF</th>
      <td>302409</td>
    </tr>
    <tr>
      <th>MB</th>
      <td>286504</td>
    </tr>
    <tr>
      <th>OLF</th>
      <td>272845</td>
    </tr>
    <tr>
      <th>MY</th>
      <td>150360</td>
    </tr>
    <tr>
      <th>P</th>
      <td>138893</td>
    </tr>
    <tr>
      <th>HY</th>
      <td>135346</td>
    </tr>
    <tr>
      <th>TH</th>
      <td>135015</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_count</span> <span class="o">=</span> <span class="n">parcellation_cell_count</span><span class="p">(</span><span class="n">section</span><span class="p">,</span><span class="s1">&#39;parcellation_structure&#39;</span><span class="p">)</span>
<span class="n">cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>parcellation_structure</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>SSp-bfd</th>
      <td>8835</td>
    </tr>
    <tr>
      <th>SSs</th>
      <td>7191</td>
    </tr>
    <tr>
      <th>CP</th>
      <td>4723</td>
    </tr>
    <tr>
      <th>DG</th>
      <td>4493</td>
    </tr>
    <tr>
      <th>VISa</th>
      <td>4447</td>
    </tr>
    <tr>
      <th>CA1</th>
      <td>3676</td>
    </tr>
    <tr>
      <th>PIR</th>
      <td>3094</td>
    </tr>
    <tr>
      <th>cst</th>
      <td>3018</td>
    </tr>
    <tr>
      <th>RSPv</th>
      <td>2852</td>
    </tr>
    <tr>
      <th>cc</th>
      <td>2834</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can filtered cells to specific subclasses and types and count within the filtered set</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_joined</span><span class="p">[</span><span class="s1">&#39;subclass&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;063 MH Tac2 Glut&#39;</span><span class="p">)</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;number of cells:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered</span><span class="p">))</span>
<span class="n">cell_count</span> <span class="o">=</span> <span class="n">parcellation_cell_count</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span><span class="s1">&#39;parcellation_structure&#39;</span><span class="p">)</span>
<span class="n">cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>number of cells: 4694
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>parcellation_structure</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>MH</th>
      <td>4404</td>
    </tr>
    <tr>
      <th>V3-unassigned</th>
      <td>133</td>
    </tr>
    <tr>
      <th>mfsbshy</th>
      <td>72</td>
    </tr>
    <tr>
      <th>brain-unassigned</th>
      <td>47</td>
    </tr>
    <tr>
      <th>chpl</th>
      <td>20</td>
    </tr>
    <tr>
      <th>LH</th>
      <td>11</td>
    </tr>
    <tr>
      <th>DG</th>
      <td>2</td>
    </tr>
    <tr>
      <th>TH-unassigned</th>
      <td>2</td>
    </tr>
    <tr>
      <th>LP</th>
      <td>1</td>
    </tr>
    <tr>
      <th>MD</th>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_joined</span><span class="p">[</span><span class="s1">&#39;subclass&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;064 LH Pou4f1 Sox1 Glut&#39;</span><span class="p">)</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;number of cells:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered</span><span class="p">))</span>
<span class="n">cell_count</span> <span class="o">=</span> <span class="n">parcellation_cell_count</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span><span class="s1">&#39;parcellation_structure&#39;</span><span class="p">)</span>
<span class="n">cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>number of cells: 1097
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>parcellation_structure</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>LH</th>
      <td>879</td>
    </tr>
    <tr>
      <th>mfsbshy</th>
      <td>110</td>
    </tr>
    <tr>
      <th>MH</th>
      <td>60</td>
    </tr>
    <tr>
      <th>PAG</th>
      <td>10</td>
    </tr>
    <tr>
      <th>TH-unassigned</th>
      <td>10</td>
    </tr>
    <tr>
      <th>brain-unassigned</th>
      <td>7</td>
    </tr>
    <tr>
      <th>PF</th>
      <td>6</td>
    </tr>
    <tr>
      <th>MD</th>
      <td>5</td>
    </tr>
    <tr>
      <th>DG</th>
      <td>4</td>
    </tr>
    <tr>
      <th>SPA</th>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_joined</span><span class="p">[</span><span class="s1">&#39;subclass&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;065 AD Serpinb7 Glut&#39;</span><span class="p">)</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;number of cells:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered</span><span class="p">))</span>
<span class="n">cell_count</span> <span class="o">=</span> <span class="n">parcellation_cell_count</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span><span class="s1">&#39;parcellation_structure&#39;</span><span class="p">)</span>
<span class="n">cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>number of cells: 1193
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>parcellation_structure</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AD</th>
      <td>1039</td>
    </tr>
    <tr>
      <th>mfsbshy</th>
      <td>84</td>
    </tr>
    <tr>
      <th>brain-unassigned</th>
      <td>34</td>
    </tr>
    <tr>
      <th>DG</th>
      <td>23</td>
    </tr>
    <tr>
      <th>LD</th>
      <td>4</td>
    </tr>
    <tr>
      <th>AM</th>
      <td>3</td>
    </tr>
    <tr>
      <th>TH-unassigned</th>
      <td>3</td>
    </tr>
    <tr>
      <th>mfbc</th>
      <td>3</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_joined</span><span class="p">[</span><span class="s1">&#39;subclass&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;007 L2/3 IT CTX Glut&#39;</span><span class="p">)</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="n">cell_joined</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;number of cells:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered</span><span class="p">))</span>
<span class="n">cell_count</span> <span class="o">=</span> <span class="n">parcellation_cell_count</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span><span class="s1">&#39;parcellation_substructure&#39;</span><span class="p">)</span>
<span class="n">cell_count</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>number of cells: 124387
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_of_cells</th>
    </tr>
    <tr>
      <th>parcellation_substructure</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>MOs2/3</th>
      <td>12695</td>
    </tr>
    <tr>
      <th>MOp2/3</th>
      <td>10638</td>
    </tr>
    <tr>
      <th>VISp2/3</th>
      <td>9802</td>
    </tr>
    <tr>
      <th>SSs2/3</th>
      <td>7326</td>
    </tr>
    <tr>
      <th>SSp-m2/3</th>
      <td>5515</td>
    </tr>
    <tr>
      <th>SSp-bfd2/3</th>
      <td>5375</td>
    </tr>
    <tr>
      <th>TEa2/3</th>
      <td>4341</td>
    </tr>
    <tr>
      <th>SSp-ul2/3</th>
      <td>3458</td>
    </tr>
    <tr>
      <th>RSPagl2/3</th>
      <td>3239</td>
    </tr>
    <tr>
      <th>ACAv2/3</th>
      <td>3111</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="resampled-ccf-template-and-parcellations">
<h2>Resampled CCF template and parcellations<a class="headerlink" href="#resampled-ccf-template-and-parcellations" title="Permalink to this headline">#</a></h2>
<p>The anatomical template, parcellation and boundary volume in resample CCF coordinates are also available for download for visualization and analysis</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">volumes</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="s1">&#39;file_listing&#39;</span><span class="p">][</span><span class="s1">&#39;MERFISH-C57BL6J-638850-CCF&#39;</span><span class="p">][</span><span class="s1">&#39;image_volumes&#39;</span><span class="p">]</span>
<span class="n">volumes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;resampled_annotation_boundary&#39;: {&#39;files&#39;: {&#39;nii.gz&#39;: {&#39;version&#39;: &#39;20230630&#39;,
    &#39;relative_path&#39;: &#39;image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_annotation_boundary.nii.gz&#39;,
    &#39;url&#39;: &#39;https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_annotation_boundary.nii.gz&#39;,
    &#39;size&#39;: 1548196}}},
 &#39;resampled_average_template&#39;: {&#39;files&#39;: {&#39;nii.gz&#39;: {&#39;version&#39;: &#39;20230630&#39;,
    &#39;relative_path&#39;: &#39;image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_average_template.nii.gz&#39;,
    &#39;url&#39;: &#39;https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_average_template.nii.gz&#39;,
    &#39;size&#39;: 117382681}}},
 &#39;resampled_annotation&#39;: {&#39;files&#39;: {&#39;nii.gz&#39;: {&#39;version&#39;: &#39;20230630&#39;,
    &#39;relative_path&#39;: &#39;image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_annotation.nii.gz&#39;,
    &#39;url&#39;: &#39;https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_annotation.nii.gz&#39;,
    &#39;size&#39;: 2115809}}}}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading resampled_average_template&quot;</span><span class="p">)</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">volumes</span><span class="p">[</span><span class="s1">&#39;resampled_average_template&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;nii.gz&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">average_template_image</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">ReadImage</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
<span class="n">average_template_array</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">GetArrayViewFromImage</span><span class="p">(</span> <span class="n">average_template_image</span> <span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading resampled_annotation&quot;</span><span class="p">)</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">volumes</span><span class="p">[</span><span class="s1">&#39;resampled_annotation&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;nii.gz&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">annotation_image</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">ReadImage</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
<span class="n">annotation_array</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">GetArrayViewFromImage</span><span class="p">(</span> <span class="n">annotation_image</span> <span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading resampled_annotation_boundary&quot;</span><span class="p">)</span>
<span class="n">rpath</span> <span class="o">=</span> <span class="n">volumes</span><span class="p">[</span><span class="s1">&#39;resampled_annotation_boundary&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;nii.gz&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="n">annotation_boundary_image</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">ReadImage</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>
<span class="n">annotation_boundary_array</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">GetArrayViewFromImage</span><span class="p">(</span> <span class="n">annotation_boundary_image</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reading resampled_average_template
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">RuntimeError</span><span class="g g-Whitespace">                              </span>Traceback (most recent call last)
<span class="nn">/var/folders/sd/7d15v2v168sd396gsdjsn8j00000gn/T/ipykernel_2668/1590720919.py</span> in <span class="ni">?</span><span class="nt">()</span>
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading resampled_average_template&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span> <span class="n">rpath</span> <span class="o">=</span> <span class="n">volumes</span><span class="p">[</span><span class="s1">&#39;resampled_average_template&#39;</span><span class="p">][</span><span class="s1">&#39;files&#39;</span><span class="p">][</span><span class="s1">&#39;nii.gz&#39;</span><span class="p">][</span><span class="s1">&#39;relative_path&#39;</span><span class="p">]</span>
<span class="g g-Whitespace">      </span><span class="mi">6</span> <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">download_base</span><span class="p">,</span> <span class="n">rpath</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">7</span> <span class="n">average_template_image</span> <span class="o">=</span> <span class="n">sitk</span><span class="o">.</span><span class="n">ReadImage</span><span class="p">(</span> <span class="n">file</span> <span class="p">)</span>

<span class="nn">~/miniconda3/envs/jupyter_book/lib/python3.9/site-packages/SimpleITK/extra.py</span> in <span class="ni">?</span><span class="nt">(fileName, outputPixelType, imageIO)</span>
<span class="g g-Whitespace">    </span><span class="mi">351</span>         <span class="n">reader</span><span class="o">.</span><span class="n">SetFileNames</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">fileName</span><span class="p">])</span>
<span class="g g-Whitespace">    </span><span class="mi">352</span> 
<span class="g g-Whitespace">    </span><span class="mi">353</span>     <span class="n">reader</span><span class="o">.</span><span class="n">SetImageIO</span><span class="p">(</span><span class="n">imageIO</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">354</span>     <span class="n">reader</span><span class="o">.</span><span class="n">SetOutputPixelType</span><span class="p">(</span><span class="n">outputPixelType</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">355</span>     <span class="k">return</span> <span class="n">reader</span><span class="o">.</span><span class="n">Execute</span><span class="p">()</span>

<span class="nn">~/miniconda3/envs/jupyter_book/lib/python3.9/site-packages/SimpleITK/SimpleITK.py</span> in <span class="ni">?</span><span class="nt">(self)</span>
<span class="g g-Whitespace">   </span><span class="mi">8434</span>         <span class="nb">type</span> <span class="n">to</span> <span class="n">be</span> <span class="n">same</span> <span class="k">as</span> <span class="n">the</span> <span class="n">file</span><span class="o">.</span> <span class="n">If</span> <span class="n">the</span> <span class="n">pixel</span> <span class="nb">type</span> <span class="ow">is</span> <span class="n">specified</span> <span class="n">then</span> <span class="n">the</span> <span class="n">itk</span><span class="p">::</span><span class="n">ConvertPixelBuffer</span> <span class="n">will</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">convert</span> <span class="n">the</span> <span class="n">pixels</span><span class="o">.</span>
<span class="g g-Whitespace">   </span><span class="mi">8435</span> 
<span class="g g-Whitespace">   </span><span class="mi">8436</span> 
<span class="g g-Whitespace">   </span><span class="mi">8437</span>         <span class="s2">&quot;&quot;&quot;</span>
<span class="ne">-&gt; </span><span class="mi">8438</span><span class="s2">         return _SimpleITK.ImageFileReader_Execute(self)</span>

<span class="ne">RuntimeError</span>: Exception thrown in SimpleITK ImageFileReader_Execute: /Users/runner/work/1/sitk/Code/IO/src/sitkImageReaderBase.cxx:97:
<span class="ne">sitk</span>::ERROR: The file &quot;https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/image_volumes/MERFISH-C57BL6J-638850-CCF/20230630/resampled_average_template.nii.gz&quot; does not exist.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to print out image information</span>
<span class="k">def</span> <span class="nf">image_info</span><span class="p">(</span> <span class="n">img</span> <span class="p">)</span> <span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;size: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">img</span><span class="o">.</span><span class="n">GetSize</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39; voxels&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;spacing: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">img</span><span class="o">.</span><span class="n">GetSpacing</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39; mm&#39;</span> <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;direction: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">img</span><span class="o">.</span><span class="n">GetDirection</span><span class="p">())</span> <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;origin: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">img</span><span class="o">.</span><span class="n">GetOrigin</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">image_info</span><span class="p">(</span><span class="n">average_template_image</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>size: (1100, 1100, 76) voxels
spacing: (0.009999999776482582, 0.009999999776482582, 0.20000000298023224) mm
direction: (-1.0, 0.0, 0.0, 0.0, -0.0, -1.0, 0.0, -1.0, 0.0)
origin: (5.476531505584717, 8.600000381469727, 6.435669422149658)
</pre></div>
</div>
</div>
</div>
<p>To enable the overlay of image section with cell coordinates we need to compute the extent the image in mm coordinates</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">size</span> <span class="o">=</span> <span class="n">average_template_image</span><span class="o">.</span><span class="n">GetSize</span><span class="p">()</span>
<span class="n">spacing</span> <span class="o">=</span> <span class="n">average_template_image</span><span class="o">.</span><span class="n">GetSpacing</span><span class="p">()</span>
<span class="n">extent</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mf">0.5</span> <span class="o">*</span> <span class="n">spacing</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">(</span><span class="n">size</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">*</span> <span class="n">spacing</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">(</span><span class="n">size</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">*</span> <span class="n">spacing</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="o">-</span><span class="mf">0.5</span> <span class="o">*</span> <span class="n">spacing</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="visualize-cells-with-parcellation-boundary-overlay">
<h2>Visualize cells with parcellation boundary overlay<a class="headerlink" href="#visualize-cells-with-parcellation-boundary-overlay" title="Permalink to this headline">#</a></h2>
<p>For example section “C57BL6J-638850.40”, we first compute the corresponding z section in the resampled image volume and visualize the resampled average template, annotation and parcellation for that section</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">zindex</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">section</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;z_reconstructed&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="mf">0.2</span><span class="p">)</span>
<span class="n">zindex</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>36
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">template_slice</span> <span class="o">=</span> <span class="n">average_template_array</span><span class="p">[</span><span class="n">zindex</span><span class="p">,:,:]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">overlay</span><span class="o">=</span><span class="n">template_slice</span><span class="p">,</span><span class="n">extent</span><span class="o">=</span><span class="n">extent</span><span class="p">)</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;resampled_average_template&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/63538ceeeb8acf60afa58678145fdde20c97363fbc4568148ecb7aa930225598.png" src="../_images/63538ceeeb8acf60afa58678145fdde20c97363fbc4568148ecb7aa930225598.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">annotation_slice</span> <span class="o">=</span> <span class="n">annotation_array</span><span class="p">[</span><span class="n">zindex</span><span class="p">,:,:]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">overlay</span><span class="o">=</span><span class="n">annotation_slice</span><span class="p">,</span><span class="n">extent</span><span class="o">=</span><span class="n">extent</span><span class="p">)</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;resampled_annotation&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b0af2898beca158a745283f6aae9e499252ca28397a53a766ca68ad4e240c5b2.png" src="../_images/b0af2898beca158a745283f6aae9e499252ca28397a53a766ca68ad4e240c5b2.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">boundary_slice</span> <span class="o">=</span> <span class="n">annotation_boundary_array</span><span class="p">[</span><span class="n">zindex</span><span class="p">,:,:]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">overlay</span><span class="o">=</span><span class="n">boundary_slice</span><span class="p">,</span><span class="n">bcmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Greys</span><span class="p">,</span><span class="n">extent</span><span class="o">=</span><span class="n">extent</span><span class="p">)</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;resampled_annotation_boundary&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b3c589506644d79ba2022c1f70ccdce43715b69bed5797881cb14b28b6d5d20e.png" src="../_images/b3c589506644d79ba2022c1f70ccdce43715b69bed5797881cb14b28b6d5d20e.png" />
</div>
</div>
<p>Using the help function, we can overlay the parcellation boundary overlay on top of the cells colorized by subclass</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_section</span><span class="p">(</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;x_reconstructed&#39;</span><span class="p">],</span> <span class="n">section</span><span class="p">[</span><span class="s1">&#39;y_reconstructed&#39;</span><span class="p">],</span> 
                        <span class="n">cc</span><span class="o">=</span><span class="n">section</span><span class="p">[</span><span class="s1">&#39;subclass_color&#39;</span><span class="p">],</span>
                        <span class="n">overlay</span><span class="o">=</span><span class="n">boundary_slice</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="n">extent</span><span class="p">,</span> 
                        <span class="n">bcmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Greys</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">1.0</span><span class="o">*</span><span class="p">(</span><span class="n">boundary_slice</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">),</span>
                        <span class="n">fig_width</span> <span class="o">=</span> <span class="mi">9</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="mi">9</span> <span class="p">)</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Subclass - Reconstructed Coordinates&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/0eb0ed9d058546fc92a1058872817c9cb5ee57bd8191c647851c477131afb14a.png" src="../_images/0eb0ed9d058546fc92a1058872817c9cb5ee57bd8191c647851c477131afb14a.png" />
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="ccf_and_parcellation_annotation_tutorial.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Allen Mouse Common Coordinate Framework (2020 version)</p>
      </div>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-in-section-reconstructed-and-ccf-coordinates-for-all-cells">Read in section, reconstructed and CCF coordinates for all cells</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-cells-in-section-reconstruced-and-ccf-coordinates">Visualize cells in section, reconstruced and CCF coordinates</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#associate-mapped-parcellation-and-parcellation-terms-to-the-merfish-data">Associate mapped parcellation and parcellation terms to the MERFISH data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-cell-parcellation-assignment-at-different-anatomical-levels">Visualize cell parcellation assignment at different anatomical levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#aggregate-cells-by-parcellation-assignment-at-different-anatomical-levels">Aggregate cells by parcellation assignment at different anatomical levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#resampled-ccf-template-and-parcellations">Resampled CCF template and parcellations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-cells-with-parcellation-boundary-overlay">Visualize cells with parcellation boundary overlay</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Allen Institute
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>